// @edgeone/pages-agent-toolkit/dist/tools/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tools/index.ts
var tools_exports = {};
__export(tools_exports, {
  buildTools: () => buildTools,
  isToolErrorResult: () => isToolErrorResult,
  toToolErrorResult: () => toToolErrorResult
});
module.exports = __toCommonJS(tools_exports);

// src/tools/websearch/http/client.ts
var import_axios = __toESM(require("axios"), 1);
var SEARCH_PRO_ENDPOINT = "https://api.wsa.cloud.tencent.com/SearchPro";
var DEFAULT_TIMEOUT_MS = 15e3;
var WSA_ERROR_HINTS = {
  InternalError: "Tencent web search internal error. Retry later.",
  InvalidParameter: "Invalid request parameter sent to the web search API.",
  RequestLimitExceeded: "Web search request rate limit exceeded. Retry later.",
  ResourceNotFound: "Web search service is not enabled for this account. Ask the main account to enable Tencent Cloud Web Search (WSA) in the Tencent Cloud console.",
  ResourceUnavailable: "Web search resource is unavailable. Check whether the account is in arrears.",
  UnauthorizedOperation: "Unauthorized web search request. Check that WSA_API_KEY is correct and active."
};
var TencentWsaError = class extends Error {
  constructor(message, opts = { code: "InternalError" }) {
    super(message);
    this.name = "TencentWsaError";
    this.code = opts.code;
    this.requestId = opts.requestId;
    this.status = opts.status;
  }
};
function isDebugEnabled() {
  const env = globalThis.process?.env?.MAKERS_AGENT_TOOLKIT_DEBUG;
  return env === "1" || env === "true";
}
function debugLog(event, data) {
  if (!isDebugEnabled()) return;
  const suffix = data ? ` ${JSON.stringify(data)}` : "";
  console.error(`[pages-agent-toolkit] ${event}${suffix}`);
}
var AxiosWsaTransport = class {
  async post(url, init) {
    const resp = await import_axios.default.request({
      method: "POST",
      url,
      data: init.body,
      headers: init.headers,
      timeout: init.timeoutMs ?? DEFAULT_TIMEOUT_MS,
      responseType: "json",
      transformRequest: [(data) => data],
      validateStatus: () => true
    });
    return { status: resp.status, data: resp.data };
  }
};
function buildAuthError() {
  throw new TencentWsaError(
    "web_search requires the WSA_API_KEY environment variable. Set it to a valid Tencent Cloud Web Search (WSA) API key before calling web_search.",
    { code: "UnauthorizedOperation" }
  );
}
function resolveApiKey() {
  const key = globalThis.process?.env?.WSA_API_KEY;
  if (typeof key !== "string" || key.trim() === "") buildAuthError();
  return key.trim();
}
function parseJson(value) {
  if (value == null) return value;
  if (typeof value === "string") {
    try {
      return JSON.parse(value);
    } catch {
      return value;
    }
  }
  return value;
}
function extractApiError(payload) {
  if (!payload || typeof payload !== "object") return null;
  const root = payload.Response && typeof payload.Response === "object" ? payload.Response : payload;
  const requestId = root.RequestId ?? root.requestId ?? payload.RequestId;
  const error = root.Error ?? root.error;
  if (error && typeof error === "object" && (error.Code ?? error.code)) {
    return {
      code: String(error.Code ?? error.code),
      message: String(error.Message ?? error.message ?? ""),
      requestId
    };
  }
  if (typeof root.Code === "string" && root.Code && root.Pages === void 0) {
    return { code: root.Code, message: String(root.Message ?? ""), requestId };
  }
  return null;
}
function throwApiError(err, status) {
  const hint = WSA_ERROR_HINTS[err.code] ?? "Web search request failed.";
  const detail = err.message ? ` (${err.message})` : "";
  const requestSuffix = err.requestId ? ` [RequestId: ${err.requestId}]` : "";
  throw new TencentWsaError(`web_search failed: ${hint}${detail}${requestSuffix}`, {
    code: err.code,
    requestId: err.requestId,
    status
  });
}
async function searchPro(query, params = {}, transport = new AxiosWsaTransport()) {
  const apiKey = resolveApiKey();
  const requestBody = { Query: query };
  const site = params.site?.trim();
  if (site) requestBody.Site = site;
  const body = JSON.stringify(requestBody);
  debugLog("websearch.wsa.request", { endpoint: SEARCH_PRO_ENDPOINT, site: site || void 0 });
  const { status, data } = await transport.post(SEARCH_PRO_ENDPOINT, {
    headers: {
      "Content-Type": "application/json; charset=UTF-8",
      Authorization: `Bearer ${apiKey}`
    },
    body
  });
  debugLog("websearch.wsa.response", { status });
  const payload = parseJson(data);
  if (status < 200 || status >= 300) {
    const apiError2 = extractApiError(payload);
    if (apiError2) throwApiError(apiError2, status);
    const text = typeof payload === "string" ? payload : JSON.stringify(payload ?? {});
    throw new TencentWsaError(`web_search failed: HTTP ${status} from the web search API. ${text}`.trim(), {
      code: "InternalError",
      status
    });
  }
  const apiError = extractApiError(payload);
  if (apiError) throwApiError(apiError, status);
  const root = payload && payload.Response && typeof payload.Response === "object" ? payload.Response : payload;
  const rawPages = Array.isArray(root?.Pages) ? root.Pages : [];
  const pages = rawPages.map((entry) => parseJson(entry)).filter((entry) => Boolean(entry) && typeof entry === "object");
  return {
    query: root?.Query,
    pages,
    version: root?.Version,
    msg: root?.Msg,
    requestId: root?.RequestId
  };
}

// src/tools/websearch/search.ts
var ENGINE_NAME = "tencent-wsa";
var SimpleSearch = class {
  constructor(transport = new AxiosWsaTransport()) {
    this.transport = transport;
  }
  async search(query, options = {}) {
    if (!query || !query.trim()) {
      throw new Error("query is required");
    }
    const maxResults = options.maxResults ?? 5;
    if (!Number.isInteger(maxResults) || maxResults <= 0) {
      throw new Error("maxResults must be a positive integer");
    }
    const { pages } = await searchPro(query.trim(), { site: options.site }, this.transport);
    return mapPages(pages, maxResults);
  }
};
function mapPages(pages, maxResults) {
  const seen = /* @__PURE__ */ new Set();
  const results = [];
  for (const page of pages) {
    const href = (page.url ?? "").trim();
    if (!href || seen.has(href)) continue;
    seen.add(href);
    results.push({
      title: (page.title ?? "").trim(),
      href,
      snippet: (page.content || page.passage || "").trim(),
      site: (page.site ?? "").trim(),
      date: (page.date ?? "").trim(),
      engine: ENGINE_NAME
    });
    if (results.length >= maxResults) break;
  }
  return results;
}

// src/tools/registry.ts
function getValue(value, key, defaultValue) {
  if (value && typeof value === "object" && key in value) return value[key];
  return defaultValue;
}
function normalizeCommandResult(result) {
  return {
    stdout: getValue(result, "stdout", ""),
    stderr: getValue(result, "stderr", ""),
    exitCode: getValue(result, "exitCode", getValue(result, "exit_code", 0))
  };
}
function normalizeEntry(entry) {
  return {
    name: getValue(entry, "name"),
    type: getValue(entry, "type"),
    path: getValue(entry, "path")
  };
}
function normalizeExecution(execution) {
  const payload = {
    results: getValue(execution, "results", []),
    logs: getValue(execution, "logs", { stdout: [], stderr: [] })
  };
  const error = getValue(execution, "error");
  if (error !== void 0 && error !== null) payload.error = error;
  return payload;
}
function normalizeBrowserResult(result) {
  const payload = { success: getValue(result, "success") };
  const data = getValue(result, "data");
  if (data !== void 0 && data !== null) payload.data = data;
  const error = getValue(result, "error");
  if (error !== void 0 && error !== null) payload.error = error;
  return payload;
}
function asRecord(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}
var BUILTIN_TOOLS = [
  {
    name: "commands",
    description: 'Use this tool when the agent needs to run a one-shot shell command in the sandbox, including builds, tests, package managers, CLI tools, asset downloads, or binary file generation. Prefer files_read or files_write for direct text file reads or writes. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot for page interaction. Prefer code_interpreter for stateful analysis or notebook-style computation. Returns stdout, stderr, and exitCode; exitCode !== 0 means the command failed or completed with an error status. Side effects: shell commands can mutate files, install packages, start processes, or delete data, so destructive commands require extra care. Input example: { "cmd": "npm test", "cwd": "/workspace", "timeout": 60 }. Common mistakes: do not use commands for long-running daemons; do not use shell cat/echo/sed to read or write a single text file when files_read or files_write fits; do not ignore non-zero exitCode. Supports optional cwd, env, and timeout in seconds.',
    inputSchema: {
      type: "object",
      properties: {
        cmd: { type: "string", description: "Shell command to execute as a one-shot sandbox process" },
        cwd: { type: "string", description: "Working directory for the command (optional)" },
        env: { type: "object", description: "Environment variables for this command only (optional)" },
        timeout: { type: "number", description: "Maximum execution time in seconds (optional)" }
      },
      required: ["cmd"]
    },
    execute: async ({ cmd, cwd, env, timeout }, sandbox) => {
      const result = await sandbox.commands.run(cmd, { cwd, env, timeout });
      return normalizeCommandResult(result);
    }
  },
  {
    name: "files_read",
    description: 'Use this tool when the agent needs to read the content of a sandbox text file without running shell commands. Prefer commands for grep/find or shell pipelines, files_list for directory listing, files_write for writing text content, and code_interpreter for executing code or data analysis. Output semantics: returns the file content as a string. Input example: { "path": "/workspace/package.json" }. Common mistakes: do not use files_read for recursive search or shell pipelines; use files_list for directories and files_write for text writes.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file path to read" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => await sandbox.files.read(path)
  },
  {
    name: "files_write",
    description: 'Use this tool when the agent needs to create or overwrite a sandbox text file with exact UTF-8/text content. Prefer commands for generated shell output, pipelines, asset downloads, binary/cache files, or decoding base64 into binary files, and files_read for reading existing text content first. Output semantics: returns { success: true }. Side effects: overwrites existing text file content or creates a new text file. Input example: { "path": "/tmp/out.txt", "content": "hello" }. Common mistakes: writes replace the whole file; content is required; do not use files_write for binary or cache assets such as .mp3, images, archives, or decoded bytes; use files_make_dir for directories and commands for complex generated pipelines.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox text file path to write" },
        content: { type: "string", description: "Exact UTF-8/text content to write; not for binary/cache assets or decoded bytes" }
      },
      required: ["path", "content"]
    },
    execute: async ({ path, content }, sandbox) => {
      if (typeof content !== "string") throw new Error("files_write requires content");
      await sandbox.files.write(path, content);
      return { success: true };
    }
  },
  {
    name: "files_list",
    description: 'Use this tool when the agent needs to list a sandbox directory without running shell commands. Prefer commands for recursive find/grep pipelines. Output semantics: returns entries with name, type, and path only. Input example: { "path": "/workspace" }. Common mistakes: use files_read for file contents; use commands for recursive or filtered searches.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox directory path to list" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      const entries = await sandbox.files.list(path);
      return entries.map((entry) => normalizeEntry(entry));
    }
  },
  {
    name: "files_exists",
    description: 'Use this tool when the agent needs to check whether a sandbox file or directory exists. Output semantics: returns a boolean. Input example: { "path": "/workspace/package.json" }. Common mistakes: files_exists only checks existence; use files_read to inspect content and files_list to inspect directories.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file or directory path to check" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => await sandbox.files.exists(path)
  },
  {
    name: "files_remove",
    description: 'Use this tool when the agent needs to remove a sandbox file or directory. Output semantics: returns { success: true }. Side effects: deletes the target path, so destructive removals require extra care. Input example: { "path": "/tmp/out.txt" }. Common mistakes: do not call files_remove for cleanup unless deletion is intended; use files_exists first when uncertain.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file or directory path to remove" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      await sandbox.files.remove(path);
      return { success: true };
    }
  },
  {
    name: "files_make_dir",
    description: 'Use this tool when the agent needs to create a sandbox directory, including nested directories supported by the sandbox. Output semantics: returns { success: true }. Side effects: creates directories on the sandbox filesystem. Input example: { "path": "/tmp/new-dir" }. Common mistakes: use files_write for file content; this tool creates directories only and does not accept make_dir or makeDir op values.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox directory path to create" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      await sandbox.files.makeDir(path);
      return { success: true };
    }
  },
  {
    name: "browser_fetch",
    description: 'Use this tool when the agent already has a specific target URL and needs a real Chromium browser inside the sandbox to navigate to a page and inspect DOM content at that URL. browser_fetch performs full browser navigation and DOM capture, not a lightweight HTTP/API fetch and not a search tool. Prefer commands with curl or code_interpreter HTTP clients for simple JSON, text, or API requests, and prefer web_search for public discovery, latest news, trending topics, or any information lookup where the exact target URL is not known. Output semantics: returns page metadata plus content truncated to 5000 characters. Side effects: navigation changes the current browser page. Input example: { "url": "https://example.com" }. Common mistakes: do not use browser_fetch for general web search or information discovery; use web_search when you do not already have a specific target URL; do not use browser_fetch for direct API fetches, screenshots, clicks, or form input.',
    inputSchema: {
      type: "object",
      properties: {
        url: { type: "string", description: "Known target URL for a real browser page load and DOM capture; not a lightweight HTTP request and not for search or discovery" }
      },
      required: ["url"]
    },
    execute: async ({ url }, sandbox) => {
      if (typeof url !== "string" || !url) throw new Error("browser_fetch requires url");
      const navigation = await sandbox.browser.goto(url);
      const content = await sandbox.browser.getContent();
      const navigationData = asRecord(getValue(navigation, "data", {}));
      const pageContent = String(getValue(content, "data", "") ?? "");
      return {
        ...navigationData,
        content: pageContent.length > 5e3 ? pageContent.slice(0, 5e3) : pageContent
      };
    }
  },
  {
    name: "browser_screenshot",
    description: 'Use this tool when the agent needs to capture a screenshot from the sandbox browser. Optionally navigates to a URL before capture. Output semantics: returns { base64Image } containing PNG data and does not create a sandbox file or return a file path; the base64Image is the final image payload and can be returned directly to the user as-is. Side effects: optional navigation changes the current browser page. Input example: { "url": "https://example.com", "fullPage": true }. Common mistakes: screenshot returns base64Image, not a file path; do not save the base64Image to a file or convert it before responding unless the user explicitly asks for a file; use browser_fetch for DOM text and browser_click or browser_type for interaction.',
    inputSchema: {
      type: "object",
      properties: {
        url: { type: "string", description: "Optional target URL to navigate before capturing the screenshot" },
        fullPage: { type: "boolean", description: "Capture full page screenshot (default: false). Screenshot output is base64Image, not a file path and does not need to be saved before returning it." }
      },
      required: []
    },
    execute: async ({ url, fullPage }, sandbox) => {
      if (typeof url === "string" && url) await sandbox.browser.goto(url);
      const result = await sandbox.browser.screenshot({ fullPage: Boolean(fullPage) });
      return { base64Image: getValue(result, "base64Image", getValue(result, "base64_image")) };
    }
  },
  {
    name: "browser_click",
    description: 'Use this tool when the agent needs to click an element in the current sandbox browser page. Output semantics: returns normalized success/data/error payloads. Side effects: clicking can change page state. Input example: { "selector": "#submit" }. Common mistakes: browser_click requires a CSS selector and depends on the current page state; use browser_fetch first if navigation or DOM capture is needed.',
    inputSchema: {
      type: "object",
      properties: {
        selector: { type: "string", description: "CSS selector of the element to click" }
      },
      required: ["selector"]
    },
    execute: async ({ selector }, sandbox) => {
      if (typeof selector !== "string" || !selector) throw new Error("browser_click requires selector");
      const result = await sandbox.browser.click(selector);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "browser_type",
    description: 'Use this tool when the agent needs to type text into an element in the current sandbox browser page. Output semantics: returns normalized success/data/error payloads. Side effects: typing can change page state. Input example: { "selector": "#name", "text": "Alice" }. Common mistakes: browser_type requires both selector and text and depends on the current page state; it does not submit forms unless typing itself triggers submission.',
    inputSchema: {
      type: "object",
      properties: {
        selector: { type: "string", description: "CSS selector of the input element to type into" },
        text: { type: "string", description: "Text to type into the selected element" }
      },
      required: ["selector", "text"]
    },
    execute: async ({ selector, text }, sandbox) => {
      if (typeof selector !== "string" || !selector || typeof text !== "string") throw new Error("browser_type requires selector and text");
      const result = await sandbox.browser.type(selector, text);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "browser_evaluate",
    description: 'Use this tool when the agent needs to evaluate JavaScript in the current sandbox browser page context. Output semantics: returns normalized success/data/error payloads. Side effects: evaluated scripts can change page state. Input example: { "script": "document.title" }. Common mistakes: browser_evaluate runs in the current page and can mutate it; use code_interpreter for sandbox computation and commands for CLI workflows.',
    inputSchema: {
      type: "object",
      properties: {
        script: { type: "string", description: "JavaScript code to evaluate in the page context" }
      },
      required: ["script"]
    },
    execute: async ({ script }, sandbox) => {
      if (typeof script !== "string" || !script) throw new Error("browser_evaluate requires script");
      const result = await sandbox.browser.evaluate(script);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "code_interpreter",
    description: 'Use this tool when the agent needs to execute code in a stateful Jupyter kernel inside the sandbox for data analysis, calculations, notebooks, and exploratory code. Variables and imports persist across calls, making it useful for iterative computation. Prefer commands for project builds, tests, package managers, and long-running shell workflows. Prefer files_read or files_write for direct text file edits. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot for real page interaction. Returns results, logs.stdout, logs.stderr, and error when execution fails. Side effects: kernel state persists across calls and can affect later executions. Input example: { "language": "python", "code": "import statistics\\nstatistics.mean([1, 2, 3])", "timeout": 30 }. Common mistakes: do not use code_interpreter for project tests or package-manager commands; do not assume a clean kernel; do not treat it as a stateless executor; use commands for CLI workflows. Supports optional timeout in seconds.',
    inputSchema: {
      type: "object",
      properties: {
        language: { type: "string", enum: ["python", "javascript", "r", "bash"], description: "Kernel language to execute: python, javascript, r, or bash" },
        code: { type: "string", description: "Code to execute in the selected stateful kernel" },
        timeout: { type: "number", description: "Maximum execution time in seconds for this code cell (optional)" }
      },
      required: ["language", "code"]
    },
    execute: async ({ language, code, timeout }, sandbox) => {
      const result = await sandbox.runCode(code, { language, timeout });
      return normalizeExecution(result);
    }
  },
  {
    // node-websearch-tool.REGISTRY.1 node-websearch-tool.SCOPE.2 node-websearch-tool.SCOPE.3
    name: "web_search",
    description: 'Use this tool when the agent needs lightweight web search results for current facts, documentation, references, or discovery. Use web_search as the first choice to find, discover, or look up information from the web, including latest news, trending topics, real-time events, documentation, references, or any query where a specific URL is not already known. It is backed by the Tencent Cloud web search (WSA) API and requires the WSA_API_KEY environment variable. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot when the agent needs real page interaction, DOM inspection, screenshots, form entry, or JavaScript execution; browser_fetch performs real Chromium navigation and DOM capture, not general search. Prefer commands or code_interpreter for direct HTTP/API requests, local computation, project builds, or sandbox file work. Returns an array of results with title, href, and snippet, plus the source site and publication date when available. Side effects: performs outbound web requests only and does not use or mutate the sandbox filesystem, shell, browser, or kernel state. Input example: { "query": "EdgeOne Pages documentation", "maxResults": 5 }. Common mistakes: do not use web_search for interacting with a specific page; do not use it for direct JSON/API fetching; do not expect browser session state, screenshots, or JavaScript execution; do not call browser_fetch first for open-ended searches or latest-news queries without a known URL. Supports an optional maxResults limit and an optional site for site-restricted search.',
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string", description: "Search query to send to the web search API. Required and must not be empty. Use this when the target URL is unknown." },
        maxResults: { type: "integer", minimum: 1, description: "Maximum number of deduplicated search results to return (optional, default 5). Must be a positive integer." },
        site: { type: "string", description: 'Optional single domain to restrict results to (site-restricted search), e.g. "zhihu.com". Omit to search the whole web.' }
      },
      required: ["query"]
    },
    execute: async ({ query, maxResults, site }) => {
      const search = new SimpleSearch();
      const options = { maxResults };
      if (site !== void 0 && site !== null) options.site = site;
      return search.search(query, options);
    }
  }
];

// src/tools/errors.ts
function isToolErrorResult(value) {
  if (!value || typeof value !== "object") return false;
  const record = value;
  if (record.isError !== true) return false;
  const error = record.error;
  return !!error && typeof error === "object" && typeof error.message === "string";
}
function toToolErrorResult(toolName, error) {
  const message = error instanceof Error ? error.message : String(error);
  const type = error instanceof Error && error.name ? error.name : "Error";
  return { isError: true, error: { tool: toolName, type, message } };
}

// src/tools/adapters/claude.ts
function wrapForClaude(meta, sandbox) {
  return {
    name: meta.name,
    description: meta.description,
    input_schema: meta.inputSchema,
    async execute(args) {
      return await meta.execute(args, sandbox);
    }
  };
}

// src/tools/adapters/openai.ts
function wrapForOpenAI(meta, sandbox) {
  return {
    type: "function",
    function: {
      name: meta.name,
      description: meta.description,
      parameters: meta.inputSchema
    },
    async execute(args) {
      return await meta.execute(args, sandbox);
    }
  };
}

// src/tools/adapters/openai-strict-schema.ts
function stripNullishArgs(args) {
  const normalized = {};
  for (const [key, value] of Object.entries(args || {})) {
    if (value !== null && value !== void 0) normalized[key] = value;
  }
  return normalized;
}
function withDescription(base, propertySchema) {
  void propertySchema;
  return base;
}
function propertyToOpenAIStrictSchema(propertySchema) {
  if (Array.isArray(propertySchema?.enum)) {
    const values = propertySchema.enum.filter((item) => typeof item === "string");
    return withDescription(
      values.length > 0 ? { type: "string", enum: values } : { type: "string" },
      propertySchema
    );
  }
  const type = Array.isArray(propertySchema?.type) ? propertySchema.type.find((item) => item !== "null") : propertySchema?.type;
  if (type === "number" || type === "integer") return withDescription({ type: "number" }, propertySchema);
  if (type === "boolean") return withDescription({ type: "boolean" }, propertySchema);
  if (type === "object") {
    return withDescription({ type: "object", additionalProperties: {} }, propertySchema);
  }
  if (type === "array") return withDescription({ type: "array", items: {} }, propertySchema);
  return withDescription({ type: "string" }, propertySchema);
}
function toNullableSchema(propertySchema) {
  return {
    anyOf: [propertyToOpenAIStrictSchema(propertySchema), { type: "null" }]
  };
}
function jsonSchemaToOpenAIStrictSchema(schema) {
  const properties = schema?.properties && typeof schema.properties === "object" ? schema.properties : {};
  const required = Array.isArray(schema?.required) ? new Set(schema.required) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(properties)) {
    shape[key] = required.has(key) ? propertyToOpenAIStrictSchema(propertySchema) : toNullableSchema(propertySchema);
  }
  return {
    type: "object",
    properties: shape,
    required: Object.keys(properties),
    additionalProperties: false
  };
}

// src/tools/adapters/openai-agents.ts
function parseToolInput(payload) {
  if (typeof payload === "string") {
    try {
      const parsed = JSON.parse(payload);
      return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : {};
    } catch {
      return {};
    }
  }
  return payload && typeof payload === "object" && !Array.isArray(payload) ? payload : {};
}
function wrapForOpenAIAgents(meta, sandbox) {
  const execute = async (args) => {
    return await meta.execute(stripNullishArgs(args), sandbox);
  };
  return {
    type: "function",
    name: meta.name,
    description: meta.description,
    strict: true,
    parameters: jsonSchemaToOpenAIStrictSchema(meta.inputSchema),
    async execute(args) {
      return await execute(args);
    },
    async invoke(contextOrInput, input) {
      const payload = input === void 0 ? contextOrInput : input;
      return await execute(parseToolInput(payload));
    },
    async needsApproval() {
      return false;
    },
    async isEnabled() {
      return true;
    }
  };
}

// src/tools/adapters/framework-helpers.ts
var import_zod = require("zod");
function stringifyToolResult(result) {
  if (typeof result === "string") return result;
  try {
    return JSON.stringify(result);
  } catch {
    return String(result);
  }
}
function stripNullishArgs2(args) {
  return Object.fromEntries(
    Object.entries(args || {}).filter(([, value]) => value !== null && value !== void 0)
  );
}
function primitiveSchemaToZod(schema) {
  const enumValues = Array.isArray(schema?.enum) ? schema.enum.filter((value) => typeof value === "string") : [];
  if (enumValues.length > 0) {
    return import_zod.z.enum(enumValues);
  }
  switch (schema?.type) {
    case "string":
      return import_zod.z.string();
    case "number":
      return import_zod.z.number();
    case "integer":
      return import_zod.z.number().int();
    case "boolean":
      return import_zod.z.boolean();
    case "array":
      return import_zod.z.array(jsonSchemaToZod(schema.items || {}));
    case "object":
      return jsonSchemaToZod(schema);
    default:
      return import_zod.z.any();
  }
}
function withDescription2(zodSchema, schema) {
  return typeof schema?.description === "string" && schema.description ? zodSchema.describe(schema.description) : zodSchema;
}
function jsonSchemaToZod(schema) {
  if (schema?.type !== "object") {
    return withDescription2(primitiveSchemaToZod(schema), schema);
  }
  if (!schema?.properties || typeof schema.properties !== "object") {
    return withDescription2(import_zod.z.record(import_zod.z.string(), import_zod.z.any()), schema);
  }
  const required = Array.isArray(schema.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, value] of Object.entries(schema.properties)) {
    const fieldSchema = withDescription2(primitiveSchemaToZod(value), value);
    shape[key] = required.has(key) ? fieldSchema : fieldSchema.nullish();
  }
  return import_zod.z.object(shape);
}
function bindFrameworkHandler(meta, sandbox) {
  return async (args) => await meta.execute(stripNullishArgs2(args), sandbox);
}
function buildLangChainTool(meta, sandbox, toolFactory) {
  const handler = bindFrameworkHandler(meta, sandbox);
  const wrappedTool = toolFactory(
    async (args) => stringifyToolResult(await handler(args)),
    {
      name: meta.name,
      description: meta.description,
      schema: jsonSchemaToZod(meta.inputSchema)
    }
  );
  return Object.assign(wrappedTool, { handler });
}
function isClassFactory(value) {
  return typeof value === "function" && /^class\s/.test(Function.prototype.toString.call(value));
}
function buildCrewAITool(meta, sandbox, baseTool) {
  const handler = bindFrameworkHandler(meta, sandbox);
  const coroutine = async (args) => stringifyToolResult(await handler(args));
  const config = {
    name: meta.name,
    description: meta.description,
    args_schema: jsonSchemaToZod(meta.inputSchema)
  };
  let crewTool;
  if (isClassFactory(baseTool)) {
    class SandboxCrewAITool extends baseTool {
      async _run(args = {}) {
        return await coroutine(args);
      }
      async _arun(args = {}) {
        return await coroutine(args);
      }
    }
    crewTool = new SandboxCrewAITool(config);
  } else if (typeof baseTool === "function") {
    const factory = baseTool;
    crewTool = factory({
      ...config,
      coroutine,
      handler
    });
    if (typeof crewTool._run !== "function") {
      crewTool._run = async (args = {}) => await coroutine(args);
    }
    if (typeof crewTool._arun !== "function") {
      crewTool._arun = async (args = {}) => await coroutine(args);
    }
  } else {
    throw new TypeError("baseTool must be a CrewAI-compatible class or factory");
  }
  return Object.assign(crewTool, { handler });
}

// src/tools/adapters/langgraph.ts
function wrapForLangGraph(meta, sandbox) {
  return buildLangChainTool(meta, sandbox, (handler, config) => ({
    ...config,
    invoke: handler
  }));
}

// src/tools/adapters/crewai.ts
var LightweightCrewAITool = class {
  constructor(config) {
    this.name = config.name;
    this.description = config.description;
    this.args_schema = config.args_schema;
  }
};
function wrapForCrewAI(meta, sandbox) {
  return buildCrewAITool(meta, sandbox, LightweightCrewAITool);
}

// src/tools/adapters/deepagents.ts
function wrapForDeepAgents(meta, sandbox) {
  return buildLangChainTool(meta, sandbox, (handler, config) => ({
    ...config,
    invoke: handler,
    call: handler
  }));
}

// src/tools/adapters/claude-mcp.ts
var import_zod2 = require("zod");
var DEFAULT_CLAUDE_MCP_SERVER_NAME = "edgeone";
function normalizeServerName(name) {
  const normalized = typeof name === "string" ? name.trim() : "";
  return normalized || DEFAULT_CLAUDE_MCP_SERVER_NAME;
}
function withDescription3(zodSchema, schema) {
  return typeof schema?.description === "string" && schema.description ? zodSchema.describe(schema.description) : zodSchema;
}
function schemaToZod(schema) {
  const enumValues = Array.isArray(schema?.enum) ? schema.enum.filter((value) => typeof value === "string") : [];
  if (enumValues.length > 0) {
    return withDescription3(import_zod2.z.enum(enumValues), schema);
  }
  const type = Array.isArray(schema?.type) ? schema.type.find((item) => item !== "null") : schema?.type;
  let result;
  switch (type) {
    case "number":
      result = import_zod2.z.number();
      break;
    case "integer":
      result = import_zod2.z.number().int();
      break;
    case "boolean":
      result = import_zod2.z.boolean();
      break;
    case "array":
      result = import_zod2.z.array(schemaToZod(schema.items || {}));
      break;
    case "object":
      result = objectSchemaToZod(schema);
      break;
    case "string":
      result = import_zod2.z.string();
      break;
    default:
      result = import_zod2.z.any();
  }
  if (Array.isArray(schema?.type) && schema.type.includes("null")) {
    result = result.nullable();
  }
  return withDescription3(result, schema);
}
function objectSchemaToZod(schema) {
  if (!schema?.properties || typeof schema.properties !== "object") {
    return import_zod2.z.record(import_zod2.z.string(), import_zod2.z.any());
  }
  const required = Array.isArray(schema?.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(schema.properties)) {
    const field = schemaToZod(propertySchema);
    shape[key] = required.has(key) ? field : field.optional();
  }
  return import_zod2.z.object(shape);
}
function jsonSchemaToClaudeInputShape(schema) {
  if (!schema || schema.type !== "object" || !schema.properties || typeof schema.properties !== "object") {
    return void 0;
  }
  const required = Array.isArray(schema.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(schema.properties)) {
    const field = schemaToZod(propertySchema);
    shape[key] = required.has(key) ? field : field.optional();
  }
  return shape;
}
function stringifyToolResult2(result) {
  if (typeof result === "string") return result;
  const json = JSON.stringify(result, null, 2);
  return typeof json === "string" ? json : String(result);
}
function toolToClaudeMcp(meta, sandbox, options) {
  if (!meta || typeof meta.name !== "string" || !meta.name || typeof meta.description !== "string" || typeof meta.execute !== "function") {
    return void 0;
  }
  const inputSchema = jsonSchemaToClaudeInputShape(meta.inputSchema);
  if (!inputSchema) return void 0;
  const tool = {
    name: meta.name,
    description: meta.description,
    inputSchema,
    async handler(args) {
      try {
        const result = await meta.execute(args, sandbox);
        if (isToolErrorResult(result)) {
          return { content: [{ type: "text", text: result.error.message }], isError: true };
        }
        return { content: [{ type: "text", text: stringifyToolResult2(result) }] };
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return { content: [{ type: "text", text: message }], isError: true };
      }
    }
  };
  if (options?.alwaysLoad ?? true) {
    tool._meta = { "anthropic/alwaysLoad": true };
  }
  return tool;
}
function buildClaudeMcpServerBundle(metadata, sandbox, name, options) {
  const serverName = normalizeServerName(name);
  const tools = metadata.map((meta) => toolToClaudeMcp(meta, sandbox, options)).filter((tool) => Boolean(tool));
  return {
    name: serverName,
    tools,
    allowedTools: tools.map((tool) => `mcp__${serverName}__${tool.name}`)
  };
}

// src/tools/index.ts
var FILE_TOOL_NAMES = [
  "files_read",
  "files_write",
  "files_list",
  "files_exists",
  "files_remove",
  "files_make_dir"
];
var BROWSER_TOOL_NAMES = [
  "browser_fetch",
  "browser_screenshot",
  "browser_click",
  "browser_type",
  "browser_evaluate"
];
function wrapForFramework(framework, meta, sandbox) {
  switch (framework) {
    case "claude-agent-sdk":
      return wrapForClaude(meta, sandbox);
    case "openai-sdk":
      return wrapForOpenAI(meta, sandbox);
    case "openai-agents-sdk":
      return wrapForOpenAIAgents(meta, sandbox);
    case "langgraph":
      return wrapForLangGraph(meta, sandbox);
    case "crewai":
      return wrapForCrewAI(meta, sandbox);
    case "deepagents":
      return wrapForDeepAgents(meta, sandbox);
    case "basic":
    default:
      return wrapForClaude(meta, sandbox);
  }
}
function isDebugEnabled2(options) {
  const env = globalThis.process?.env?.MAKERS_AGENT_TOOLKIT_DEBUG;
  return options?.debug === true || env === "1" || env === "true";
}
function sanitize(value) {
  if (Array.isArray(value)) return value.map(sanitize);
  if (!value || typeof value !== "object") return value;
  return Object.fromEntries(Object.entries(value).map(([key, item]) => {
    const lowered = key.toLowerCase();
    if (lowered.includes("token") || lowered.includes("auth") || lowered.includes("password") || lowered.includes("secret") || lowered.includes("key")) {
      return [key, "[redacted]"];
    }
    if (key === "base64Image" && typeof item === "string") {
      return ["base64ImageLength", item.length];
    }
    return [key, sanitize(item)];
  }));
}
function summarizeResult(result) {
  if (result && typeof result === "object" && typeof result.base64Image === "string") {
    return { base64ImageLength: result.base64Image.length, savedToFile: false };
  }
  return sanitize(result);
}
function debugLog2(enabled, event, data) {
  if (!enabled) return;
  const suffix = data ? ` ${JSON.stringify(sanitize(data))}` : "";
  console.error(`[pages-agent-toolkit] ${event}${suffix}`);
}
function webSearchDebugFields(toolName, args, result) {
  if (toolName !== "web_search" || !args || typeof args !== "object") return {};
  const record = args;
  const fields = {};
  if (typeof record.query === "string") fields.query = record.query;
  if (typeof record.site === "string" && record.site) fields.site = record.site;
  if (record.maxResults !== void 0 && record.maxResults !== null) fields.maxResults = record.maxResults;
  if (Array.isArray(result)) fields.resultCount = result.length;
  return fields;
}
function stripWebSearchEngine(toolName, result) {
  if (toolName !== "web_search" || !Array.isArray(result)) return result;
  return result.map((item) => {
    if (!item || typeof item !== "object" || Array.isArray(item)) return item;
    const { engine: _engine, ...rest } = item;
    return rest;
  });
}
function withDebug(meta, enabled) {
  if (!enabled) {
    if (meta.name === "web_search") {
      return {
        ...meta,
        async execute(args, sandbox) {
          const result = await meta.execute(args, sandbox);
          return stripWebSearchEngine(meta.name, result);
        }
      };
    }
    return meta;
  }
  return {
    ...meta,
    async execute(args, sandbox) {
      const operation = meta.name;
      const start = Date.now();
      const preFields = webSearchDebugFields(meta.name, args);
      debugLog2(true, "tool.execute.start", { tool: meta.name, operation, ...preFields, args });
      try {
        const result = await meta.execute(args, sandbox);
        const postFields = webSearchDebugFields(meta.name, args, result);
        debugLog2(true, "tool.execute.success", {
          tool: meta.name,
          operation,
          ...postFields,
          durationMs: Date.now() - start,
          result: summarizeResult(result)
        });
        return stripWebSearchEngine(meta.name, result);
      } catch (error) {
        debugLog2(true, "tool.execute.error", {
          tool: meta.name,
          operation,
          ...preFields,
          durationMs: Date.now() - start,
          error: error?.message || String(error)
        });
        throw error;
      }
    }
  };
}
function withErrorHandling(meta) {
  return {
    ...meta,
    async execute(args, sandbox) {
      try {
        return await meta.execute(args, sandbox);
      } catch (error) {
        return toToolErrorResult(meta.name, error);
      }
    }
  };
}
function pickMetadata(metadata, names) {
  if (names === void 0) return metadata;
  const requestedNames = typeof names === "string" ? [names] : names;
  const seen = /* @__PURE__ */ new Set();
  const selected = [];
  for (const name of requestedNames) {
    if (seen.has(name)) continue;
    const meta = metadata.find((item) => item.name === name);
    if (!meta) throw new Error(`Unknown tool name: ${name}`);
    seen.add(name);
    selected.push(meta);
  }
  return selected;
}
function buildTools(framework, sandbox, options) {
  const fw = framework || "claude-agent-sdk";
  const toolMap = {};
  const debugEnabled = isDebugEnabled2(options);
  const metadata = BUILTIN_TOOLS.map((meta) => withErrorHandling(withDebug(meta, debugEnabled)));
  for (const meta of metadata) {
    toolMap[meta.name] = wrapForFramework(fw, meta, sandbox);
  }
  const pickTools = (names) => names.map((name) => toolMap[name]).filter(Boolean);
  return {
    commands: toolMap.commands,
    files_read: toolMap.files_read,
    files_write: toolMap.files_write,
    files_list: toolMap.files_list,
    files_exists: toolMap.files_exists,
    files_remove: toolMap.files_remove,
    files_make_dir: toolMap.files_make_dir,
    browser_fetch: toolMap.browser_fetch,
    browser_screenshot: toolMap.browser_screenshot,
    browser_click: toolMap.browser_click,
    browser_type: toolMap.browser_type,
    browser_evaluate: toolMap.browser_evaluate,
    code_interpreter: toolMap.code_interpreter,
    // node-websearch-tool.REGISTRY.2
    web_search: toolMap.web_search,
    all: () => Object.values(toolMap),
    get: (name) => toolMap[name],
    files: () => pickTools(FILE_TOOL_NAMES),
    browser: () => pickTools(BROWSER_TOOL_NAMES),
    // node-websearch-tool.COMPATIBILITY.3
    toLangChainTools: (toolFactory, names) => pickMetadata(metadata, names).map((meta) => buildLangChainTool(meta, sandbox, toolFactory)),
    // node-websearch-tool.COMPATIBILITY.4
    toCrewAITools: (baseTool, names) => pickMetadata(metadata, names).map((meta) => buildCrewAITool(meta, sandbox, baseTool)),
    toClaudeMcpServer: (name, options2) => buildClaudeMcpServerBundle(metadata, sandbox, name, options2)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildTools,
  isToolErrorResult,
  toToolErrorResult
});
