// @edgeone/is-spa/lib/constants.js
/**
 * 已知的前端路由库列表
 */
const ROUTER_LIBRARIES = [
  // Vue
  'vue-router',
  '@vue/cli-plugin-router',
  // React
  'react-router-dom',
  'react-router',
  '@tanstack/react-router',
  'wouter',
  // Angular
  '@angular/router',
  // Svelte（纯客户端路由库，用了就是 SPA）
  'svelte-routing',
  'svelte-spa-router',
  // 通用
  'page',
  'navigo',
  'universal-router',
]

/**
 * 入口文件中 router 使用的正则模式
 */
const ROUTER_IMPORT_PATTERNS = [
  // Vue: import router from './router'
  /import\s+\w+\s+from\s+['"]\.?\/?router['"]/,
  // Vue: import { createRouter } from 'vue-router'
  /import\s+\{[^}]*createRouter[^}]*\}\s+from\s+['"]vue-router['"]/,
  // Vue: .use(router)
  /\.use\(\s*router\s*\)/,
  // React: import { BrowserRouter } from 'react-router-dom'
  /import\s+\{[^}]*(?:BrowserRouter|HashRouter|MemoryRouter|Router)[^}]*\}\s+from\s+['"]react-router/,
  // React: <BrowserRouter> / <HashRouter>
  /<(?:BrowserRouter|HashRouter|Router)/,
  // Angular: RouterModule
  /RouterModule/,
  // 通用: createRouter / createWebHistory / createWebHashHistory
  /create(?:Web)?(?:History|Hash|Memory)/,
]

/**
 * 框架快速检测规则
 *
 * impliedType: 框架的默认判定类型
 * configCheck: 可选，需要读取配置文件进一步判断的框架
 *
 * 设计原则：
 *   判定依据是"构建产物是否只有单个 index.html"（即是否需要回源兜底）。
 *
 *   - Next.js / Gatsby：始终 SSG，每个路由生成独立 HTML → 不需要兜底
 *   - Nuxt：不管 ssr: true/false，nuxt generate 都为每个路由生成 HTML → 不需要兜底
 *   - SvelteKit：默认 SSG，但 fallback + ssr: false 模式产物只有单个 index.html → 需要兜底
 *   - Remix：SSR 模式有服务端处理 → 不需要兜底
 *   - CRA / Angular CLI：产物只有单个 index.html → 需要兜底
 */
const FRAMEWORK_CONVENTIONS = [
  // === 始终不需要兜底的框架 ===
  {
    name: 'Next.js',
    detect: { dependency: 'next' },
    impliedType: 'ssg',
    note: 'Next.js 使用 SSR/SSG，每个路由有对应 HTML，不需要回源兜底',
  },
  {
    name: 'Nuxt',
    detect: { dependency: 'nuxt' },
    impliedType: 'ssg',
    note: 'Nuxt 构建工具自动为每个路由生成 HTML（不管 ssr 配置），不需要回源兜底',
  },
  {
    name: 'Gatsby',
    detect: { dependency: 'gatsby' },
    impliedType: 'ssg',
    note: 'Gatsby 是 SSG 框架，每个路由生成独立 HTML，不需要回源兜底',
  },
  {
    name: 'Remix',
    detect: { dependency: '@remix-run/react' },
    impliedType: 'ssg',
    note: 'Remix 使用 SSR，每个路由有服务端处理，不需要回源兜底',
    configCheck: {
      configFiles: ['vite.config.js', 'vite.config.ts', 'vite.config.mjs', 'vite.config.mts'],
      // remix 插件中 ssr: false → 纯 SPA，构建产物只有单个 index.html
      spaPatterns: [/ssr\s*:\s*false/],
      spaNote: 'Remix 配置了 ssr: false，构建产物只有单个 index.html，需要回源兜底',
    },
  },

  {
    name: 'React Router (Framework)',
    detect: { dependency: '@react-router/dev' },
    impliedType: 'ssg',
    note: 'React Router v7 Framework 模式，默认 SSR，不需要回源兜底',
    configCheck: {
      configFiles: ['react-router.config.ts', 'react-router.config.js'],
      spaPatterns: [/ssr\s*:\s*false/],
      spaNote: 'React Router Framework 配置了 ssr: false，构建产物只有单个 index.html，需要回源兜底',
    },
  },

  // === 需要读取配置判断的框架 ===
  {
    name: 'SvelteKit',
    detect: { dependency: '@sveltejs/kit' },
    impliedType: 'ssg',
    note: 'SvelteKit 默认 SSG/SSR，每个路由生成独立 HTML，不需要回源兜底',
    configCheck: {
      configFiles: ['svelte.config.js', 'svelte.config.ts'],
      // fallback: 'xxx.html' 表示未预渲染的路由回退到单个 HTML → 纯 SPA
      spaPatterns: [/fallback\s*:\s*['"][^'"]+\.html['"]/],
      spaNote: 'SvelteKit 配置了 adapter-static fallback，构建产物只有单个 index.html，需要回源兜底',
    },
  },

  // === 始终需要兜底的框架 ===
  {
    name: 'Create React App',
    detect: { dependency: 'react-scripts' },
    impliedType: 'spa',
    note: 'CRA 构建产物只有单个 index.html，需要回源兜底',
  },
  {
    name: 'Angular CLI',
    detect: { dependency: '@angular/cli' },
    impliedType: 'spa',
    note: 'Angular CLI 构建产物只有单个 index.html，需要回源兜底',
  },
]

/**
 * 构建工具配置文件名
 */
const BUILD_CONFIG_FILES = {
  'vue-cli': ['vue.config.js', 'vue.config.ts'],
  vite: ['vite.config.js', 'vite.config.ts', 'vite.config.mjs', 'vite.config.mts'],
  webpack: ['webpack.config.js', 'webpack.config.ts'],
}

module.exports = {
  ROUTER_LIBRARIES,
  ROUTER_IMPORT_PATTERNS,
  FRAMEWORK_CONVENTIONS,
  BUILD_CONFIG_FILES,
}
