// @edgeone/is-spa/lib/index.js
const fs = require('fs')
const path = require('path')
const { detectRouter } = require('./detectors/router')
const { detectFramework } = require('./detectors/framework')
const { detectEntries } = require('./detectors/entries')
const { detectBuildOutput } = require('./detectors/build-output')

/**
 * 检测项目是 SPA、MPA、Hybrid 还是 SSG
 *
 * @param {string} projectRoot - 项目根目录的绝对路径
 * @returns {DetectionResult} 检测结果
 *
 * @typedef {Object} DetectionResult
 * @property {'spa'|'mpa'|'hybrid'|'ssg'|'unknown'} type
 * @property {'high'|'medium'|'low'} confidence
 * @property {string} reason
 * @property {Object} details
 */
function detect(projectRoot) {
  const resolvedRoot = path.resolve(projectRoot)

  // 1. 检测各维度
  const router = detectRouter(resolvedRoot)
  const framework = detectFramework(resolvedRoot)
  const entries = detectEntries(resolvedRoot)

  // 更新 router 的 hasRouterUsageInEntry
  if (entries.entryFiles && entries.entryFiles.length > 0) {
    router.hasRouterUsageInEntry = entries.entryFiles.some((e) => e.usesRouter)
  }

  const details = { router, entries, framework }

  // 2. 框架快速路径
  if (framework.name && framework.impliedType) {
    // SSG/SSR 框架：每个路由有独立 HTML，不需要回源兜底
    if (framework.impliedType === 'ssg') {
      return buildResult('ssg', 'high', details,
        `检测到 ${framework.name} 框架：${framework.note}`)
    }

    // 已知 SPA 框架，但如果有多入口且都不用 router，可能被改造成 MPA
    if (entries.totalEntryCount > 1 && entries.routerEntryCount === 0) {
      return buildResult('mpa', 'medium', details,
        `检测到 ${framework.name} 框架，但有 ${entries.totalEntryCount} 个入口且均未使用前端路由，已被改造为 MPA`)
    }

    if (entries.totalEntryCount > 1 && entries.routerEntryCount > 0 &&
        entries.routerEntryCount < entries.totalEntryCount) {
      return buildResult('hybrid', 'high', details,
        `检测到 ${framework.name} 框架，${entries.totalEntryCount} 个入口中 ${entries.routerEntryCount} 个使用路由，属于 Hybrid 架构`)
    }

    return buildResult('spa', 'high', details,
      `检测到 ${framework.name} 框架：${framework.note}`)
  }

  // 3. 基于入口数量 + 路由使用的综合判定
  const { totalEntryCount, routerEntryCount } = entries
  const nonRouterEntries = totalEntryCount - routerEntryCount
  const hasRouter = router.hasRouterDependency || router.hasRouterDirectory || router.hasRouterUsageInEntry

  // 无入口文件
  if (totalEntryCount === 0) {
    return buildResult('unknown', 'low', details,
      '未找到入口文件，无法判断项目类型')
  }

  // 单入口
  if (totalEntryCount === 1) {
    if (hasRouter) {
      return buildResult('spa', 'high', details,
        `单一入口${router.routerLibrary ? '，使用 ' + router.routerLibrary : ''} 管理前端路由，典型 SPA 架构`)
    }
    return buildResult('spa', 'medium', details,
      '单一入口但未检测到前端路由库，可能是简单单页应用或无路由需求的 SPA')
  }

  // 多入口
  if (routerEntryCount >= 1 && nonRouterEntries >= 1) {
    // 有 router 的入口 + 无 router 的入口 → Hybrid
    const routerNames = entries.entryFiles
      .filter((e) => e.usesRouter)
      .map((e) => e.name)
    const staticNames = entries.entryFiles
      .filter((e) => !e.usesRouter)
      .map((e) => e.name)

    return buildResult('hybrid', 'high', details,
      `${totalEntryCount} 个入口中，[${routerNames.join(', ')}] 使用前端路由（SPA 主体），[${staticNames.join(', ')}] 为独立页面，属于 Hybrid 架构`)
  }

  if (routerEntryCount === 0) {
    // 所有入口都没用 router → MPA
    return buildResult('mpa', 'high', details,
      `${totalEntryCount} 个独立入口，均未使用前端路由，典型 MPA 架构`)
  }

  if (routerEntryCount === totalEntryCount) {
    // 所有入口都用了 router → 多个独立 SPA（微前端）
    return buildResult('mpa', 'medium', details,
      `${totalEntryCount} 个入口均使用前端路由，可能是微前端架构（每个页面是独立 SPA）`)
  }

  return buildResult('unknown', 'low', details, '无法确定项目类型')
}

function buildResult(type, confidence, details, reason) {
  return { type, confidence, details, reason }
}

/**
 * 简化版：只返回 boolean（是否需要回源兜底到 index.html）
 * SPA 和 Hybrid 需要兜底（主体是 SPA），SSG 和 MPA 不需要
 */
function isSpa(projectRoot) {
  const result = detect(projectRoot)
  return result.type === 'spa' || result.type === 'hybrid'
}

/**
 * 判断目录更可能是源码项目还是构建产物
 */
function isSourceProject(dir) {
  const resolved = path.resolve(dir)
  const pkgPath = path.join(resolved, 'package.json')
  if (fs.existsSync(pkgPath)) {
    try {
      const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'))
      if (pkg.dependencies || pkg.devDependencies) return true
    } catch (e) {}
  }
  if (fs.existsSync(path.join(resolved, 'src'))) return true
  return false
}

/**
 * 自动检测：智能选择源码检测或构建产物检测
 * 先用判断出的模式检测，如果结果不明确则尝试另一种模式
 *
 * @param {string} dir - 目录路径（源码目录或构建产物目录均可）
 * @returns {DetectionResult & { mode: 'source' | 'dist' }}
 */
function autoDetect(dir) {
  const resolvedDir = path.resolve(dir)
  const isSource = isSourceProject(resolvedDir)
  const primaryMode = isSource ? 'source' : 'dist'
  const primaryResult = isSource ? detect(resolvedDir) : detectBuildOutput(resolvedDir)

  if (primaryResult.type !== 'unknown' || primaryResult.confidence !== 'low') {
    return { ...primaryResult, mode: primaryMode }
  }

  // 主结果不明确，尝试另一种模式
  const fallbackMode = isSource ? 'dist' : 'source'
  const fallbackResult = isSource ? detectBuildOutput(resolvedDir) : detect(resolvedDir)

  // 回退结果必须置信度为 high 才采纳，避免误判
  if (fallbackResult.type !== 'unknown' && fallbackResult.confidence === 'high') {
    return { ...fallbackResult, mode: fallbackMode }
  }

  return { ...primaryResult, mode: primaryMode }
}

/**
 * 判断是否需要 Nginx 回源兜底到 index.html
 *
 * @param {string} dir - 源码目录或构建产物目录
 * @param {Object} [options] - 可选配置
 * @param {'auto'|'source'|'dist'} [options.mode='auto'] - 检测模式
 *   - 'auto': 自动判断是源码还是构建产物（默认）
 *   - 'source': 强制使用源码检测（分析 package.json、配置文件）
 *   - 'dist': 强制使用构建产物检测（分析 HTML 和 JS 文件）
 * @returns {boolean}
 */
function needsFallback(dir, options) {
  const mode = (options && options.mode) || 'auto'
  let result

  switch (mode) {
    case 'source':
      result = detect(dir)
      break
    case 'dist':
      result = detectBuildOutput(dir)
      break
    case 'auto':
    default:
      result = autoDetect(dir)
      break
  }

  return result.type === 'spa' || result.type === 'hybrid'
}

module.exports = { detect, isSpa, needsFallback, detectBuildOutput, autoDetect }
