// @edgeone/is-spa/lib/detectors/build-output.js
const fs = require('fs')
const path = require('path')

/**
 * 在 JS 文件中搜索路由相关关键词的正则
 * 这些关键词是各路由库打包后仍会保留的特征字符串
 */
const ROUTER_KEYWORDS = [
  // History API（所有 History 模式路由库的核心）
  'pushState',
  'popstate',
  // Vue Router
  'createWebHistory',
  'createWebHashHistory',
  'RouterView',
  'router-view',
  'router-link',
  // React Router
  'BrowserRouter',
  'HashRouter',
  'createBrowserRouter',
  // Angular Router
  'RouterModule',
  'routerLink',
  // 通用特征
  'onpopstate',
]

/**
 * 检测构建产物目录，判断项目类型
 *
 * @param {string} distDir - 构建产物目录的绝对路径
 * @returns {DetectionResult} 检测结果
 */
function detectBuildOutput(distDir) {
  const resolvedDir = path.resolve(distDir)

  if (!fs.existsSync(resolvedDir) || !fs.statSync(resolvedDir).isDirectory()) {
    return buildResult('unknown', 'low', {}, '目标路径不存在或不是目录')
  }

  // 1. 检查根目录是否有 index.html
  //    兜底配置是 try_files $uri $uri/ /index.html
  //    如果根目录没有 index.html，不存在"回源首页"这件事
  const hasRootIndex = fs.existsSync(path.join(resolvedDir, 'index.html'))

  if (!hasRootIndex) {
    return buildResult('unknown', 'low', {}, '根目录没有 index.html，不需要回源兜底配置')
  }

  // 2. 收集根目录及直接子目录的 HTML 文件
  const htmlFiles = findFiles(resolvedDir, '.html', 2)
  if (htmlFiles.length === 0) {
    return buildResult('unknown', 'low', {}, '未找到 HTML 文件')
  }

  // 3. 分析每个 HTML 文件
  const htmlAnalysis = htmlFiles.map(filePath => analyzeHtml(filePath, resolvedDir))

  // 4. 收集所有 JS 文件（深度 3：覆盖 _app/immutable/chunks/、static/js/ 等深层结构）
  const jsFiles = findFiles(resolvedDir, '.js', 3)

  // 4. 检测哪些 JS 文件包含路由代码
  const routerJsFiles = findRouterJsFiles(jsFiles)

  // 5. 关联：哪些 HTML 引用了包含路由代码的 JS
  //    区分 "共享 JS"（被多个 HTML 引用）和 "页面独立 JS"（只被一个 HTML 引用）
  //    如果路由代码只出现在共享 JS 中，需要进一步看页面独立 JS 来判断
  const jsRefCount = {}  // 每个 JS 文件被多少个 HTML 引用
  for (const html of htmlAnalysis) {
    for (const src of html.scriptSrcs) {
      const basename = path.basename(src)
      jsRefCount[basename] = (jsRefCount[basename] || 0) + 1
    }
  }

  for (const html of htmlAnalysis) {
    const htmlJsBasenames = html.scriptSrcs.map(s => path.basename(s))

    // 页面独立 JS = 只被这一个 HTML 引用的 JS
    const pageSpecificJs = htmlJsBasenames.filter(js => jsRefCount[js] === 1)
    // 共享 JS = 被多个 HTML 引用的 JS
    const sharedJs = htmlJsBasenames.filter(js => jsRefCount[js] > 1)

    // 检查页面独立 JS 是否包含路由代码
    const pageJsHasRouter = pageSpecificJs.some(js =>
      routerJsFiles.some(f => path.basename(f) === js)
    )
    // 检查共享 JS 是否包含路由代码
    const sharedJsHasRouter = sharedJs.some(js =>
      routerJsFiles.some(f => path.basename(f) === js)
    )

    // 如果只有 1 个应用 HTML，共享/独立不重要，有路由就算
    // 如果有多个 HTML，优先看页面独立 JS；如果没有独立 JS，再看共享 JS
    html.hasRouterJs = pageJsHasRouter || (pageSpecificJs.length === 0 && sharedJsHasRouter)
    html.hasRouterInPageJs = pageJsHasRouter
    html.hasRouterInSharedJs = sharedJsHasRouter
  }

  // 6. 过滤
  //    - 纯静态 HTML（无 JS 引用）→ 来自 public/static 目录的原样拷贝
  //    - 框架辅助 HTML（200.html、404.html）→ Nuxt/SvelteKit 自动生成的 fallback 文件
  const FRAMEWORK_HELPER_FILES = ['200.html', '404.html']
  const appHtmlFiles = htmlAnalysis.filter(h =>
    h.scriptSrcs.length > 0 && !FRAMEWORK_HELPER_FILES.includes(path.basename(h.relativePath))
  )
  const staticHtmlFiles = htmlAnalysis.filter(h => h.scriptSrcs.length === 0)
  const helperHtmlFiles = htmlAnalysis.filter(h =>
    h.scriptSrcs.length > 0 && FRAMEWORK_HELPER_FILES.includes(path.basename(h.relativePath))
  )

  const details = {
    totalHtmlFiles: htmlFiles.length,
    appHtmlFiles: appHtmlFiles.map(h => ({
      name: h.relativePath,
      hasRouterJs: h.hasRouterJs,
      bodyEmpty: h.bodyEmpty,
      scriptCount: h.scriptSrcs.length,
    })),
    staticHtmlFiles: staticHtmlFiles.map(h => h.relativePath),
    routerJsFiles: routerJsFiles.map(f => path.relative(resolvedDir, f)),
  }

  // 7. 判定逻辑
  if (appHtmlFiles.length === 0) {
    return buildResult('unknown', 'low', details, '所有 HTML 都是纯静态页面（无 JS 引用）')
  }

  const routerHtmlCount = appHtmlFiles.filter(h => h.hasRouterJs).length
  const nonRouterHtmlCount = appHtmlFiles.filter(h => !h.hasRouterJs).length

  // 只有 1 个应用 HTML
  if (appHtmlFiles.length === 1) {
    if (routerHtmlCount === 1) {
      return buildResult('spa', 'high', details,
        `构建产物只有 1 个应用 HTML（${appHtmlFiles[0].relativePath}），且 JS 中包含路由代码，典型 SPA`)
    }
    // 有路由代码存在（可能在其他 chunk 中），但仍只有 1 个 HTML → 大概率 SPA
    if (routerJsFiles.length > 0) {
      return buildResult('spa', 'medium', details,
        `构建产物只有 1 个应用 HTML（${appHtmlFiles[0].relativePath}），JS 中存在路由代码`)
    }
    // 无路由代码 → 可能是纯静态页面，无法确定
    return buildResult('unknown', 'low', details,
      `构建产物只有 1 个应用 HTML（${appHtmlFiles[0].relativePath}），未检测到路由代码，可能是纯静态页面`)
  }

  // 多个应用 HTML — 检查是否为 SSG
  const allShareSameJs = checkSameJsBundles(appHtmlFiles)
  const allHaveContent = appHtmlFiles.every(h => !h.bodyEmpty)
  const allEmpty = appHtmlFiles.every(h => h.bodyEmpty)
  const hasRouterAnywhere = routerJsFiles.length > 0

  // SSG 判定（优先级最高）：多个 HTML 有预渲染内容 + 共享 JS
  if (appHtmlFiles.length > 1 && allHaveContent && allShareSameJs) {
    return buildResult('ssg', 'high', details,
      `${appHtmlFiles.length} 个应用 HTML 共享同一组 JS bundle，且均有预渲染内容，典型 SSG`)
  }

  // SSG 判定（宽松）：多个 HTML 有预渲染内容（即使 JS 不完全共享，如 SvelteKit 的按页 chunk）
  if (appHtmlFiles.length > 1 && allHaveContent && hasRouterAnywhere) {
    return buildResult('ssg', 'medium', details,
      `${appHtmlFiles.length} 个应用 HTML 均有预渲染内容且存在路由代码，可能是 SSG`)
  }

  if (allShareSameJs && allEmpty) {
    // Nuxt ssr:false 的 generate 产物：每个路由一个空壳 HTML，引用同一组 JS
    return buildResult('ssg', 'medium', details,
      `${appHtmlFiles.length} 个应用 HTML 共享同一组 JS bundle，body 均为空壳（可能是 Nuxt ssr:false 的 generate 产物），不需要兜底`)
  }

  // 多个 HTML，部分引用了路由 JS → Hybrid
  if (routerHtmlCount >= 1 && nonRouterHtmlCount >= 1) {
    const routerNames = appHtmlFiles.filter(h => h.hasRouterJs).map(h => h.relativePath)
    const staticNames = appHtmlFiles.filter(h => !h.hasRouterJs).map(h => h.relativePath)
    return buildResult('hybrid', 'high', details,
      `${appHtmlFiles.length} 个应用 HTML 中，[${routerNames.join(', ')}] 包含路由代码（SPA 主体），[${staticNames.join(', ')}] 为独立页面`)
  }

  // 多个 HTML，全部不含路由 → MPA
  if (routerHtmlCount === 0) {
    return buildResult('mpa', 'high', details,
      `${appHtmlFiles.length} 个应用 HTML，均不包含路由代码，典型 MPA`)
  }

  // 多个 HTML，全部含路由 → SSG 或微前端
  if (routerHtmlCount === appHtmlFiles.length) {
    if (allShareSameJs) {
      return buildResult('ssg', 'high', details,
        `${appHtmlFiles.length} 个应用 HTML 均包含路由代码且共享同一组 JS，典型 SSG`)
    }
    return buildResult('mpa', 'medium', details,
      `${appHtmlFiles.length} 个应用 HTML 均包含路由代码但 JS 各自独立，可能是微前端架构`)
  }

  return buildResult('unknown', 'low', details, '无法确定构建产物类型')
}

/**
 * 分析单个 HTML 文件
 */
function analyzeHtml(filePath, rootDir) {
  const content = fs.readFileSync(filePath, 'utf-8')
  const relativePath = path.relative(rootDir, filePath)

  // 提取 JS 引用：<script src="...">, <script type="module" src="...">, <link rel="modulepreload" href="...">
  const scriptSrcs = []
  // <script src="...">
  const scriptRegex = /<script[^>]*\bsrc=["']([^"']+\.js)["'][^>]*>/gi
  let match
  while ((match = scriptRegex.exec(content)) !== null) {
    scriptSrcs.push(match[1])
  }
  // <link rel="modulepreload" href="..."> (SvelteKit / Vite)
  const preloadRegex = /<link[^>]*\brel=["']modulepreload["'][^>]*\bhref=["']([^"']+\.js)["'][^>]*>/gi
  while ((match = preloadRegex.exec(content)) !== null) {
    scriptSrcs.push(match[1])
  }
  // 反向：<link href="..." rel="modulepreload">
  const preloadRegex2 = /<link[^>]*\bhref=["']([^"']+\.js)["'][^>]*\brel=["']modulepreload["'][^>]*>/gi
  while ((match = preloadRegex2.exec(content)) !== null) {
    if (!scriptSrcs.includes(match[1])) scriptSrcs.push(match[1])
  }

  // 检查 body 是否基本为空（只有一个空挂载点）
  const bodyMatch = content.match(/<body[^>]*>([\s\S]*?)<\/body>/i)
  let bodyEmpty = false
  if (bodyMatch) {
    // 去掉 script 标签和空白，看剩余内容长度
    const bodyContent = bodyMatch[1]
      .replace(/<script[\s\S]*?<\/script>/gi, '')
      .replace(/<noscript[\s\S]*?<\/noscript>/gi, '')
      .replace(/<!--[\s\S]*?-->/g, '')
      .replace(/<div[^>]*>\s*<\/div>/gi, '')  // 空 div
      .replace(/\s+/g, '')
    bodyEmpty = bodyContent.length < 50  // 少于 50 字符视为空壳
  }

  return { filePath, relativePath, scriptSrcs, bodyEmpty }
}

/**
 * 在 JS 文件中搜索路由相关关键词
 */
function findRouterJsFiles(jsFiles) {
  const result = []
  for (const jsFile of jsFiles) {
    try {
      const content = fs.readFileSync(jsFile, 'utf-8')
      const hasRouter = ROUTER_KEYWORDS.some(keyword => content.includes(keyword))
      if (hasRouter) {
        result.push(jsFile)
      }
    } catch (e) {
      // 读取失败跳过
    }
  }
  return result
}

/**
 * 检查多个 HTML 是否引用了同一组 JS bundle
 */
function checkSameJsBundles(htmlAnalysis) {
  if (htmlAnalysis.length <= 1) return true

  // 取每个 HTML 的 JS 引用列表，去除路径前缀只比较文件名
  const jsSets = htmlAnalysis.map(h =>
    h.scriptSrcs.map(s => path.basename(s)).sort().join(',')
  )

  // 检查是否有大部分共享（允许个别页面有额外的 chunk）
  // 策略：取最短的 JS 列表，检查其他所有 HTML 是否包含这些 JS
  const basenames = htmlAnalysis.map(h => h.scriptSrcs.map(s => path.basename(s)))
  const shortest = basenames.reduce((a, b) => a.length <= b.length ? a : b)

  // 如果最短集合中的所有 JS 都出现在每个 HTML 中 → 视为共享
  return htmlAnalysis.every(h => {
    const htmlBasenames = h.scriptSrcs.map(s => path.basename(s))
    return shortest.every(js => htmlBasenames.includes(js))
  })
}

/**
 * 递归查找指定扩展名的文件
 */
function findFiles(dir, ext, maxDepth = 2, currentDepth = 0) {
  const results = []
  if (currentDepth > maxDepth) return results
  if (!fs.existsSync(dir)) return results

  try {
    const items = fs.readdirSync(dir)
    for (const item of items) {
      if (item === 'node_modules' || item === '.git' || item === '.edgeone' || item === '.vercel' || item === '.netlify') continue
      const fullPath = path.join(dir, item)
      try {
        const stat = fs.statSync(fullPath)
        if (stat.isFile() && item.endsWith(ext)) {
          results.push(fullPath)
        } else if (stat.isDirectory() && currentDepth < maxDepth) {
          results.push(...findFiles(fullPath, ext, maxDepth, currentDepth + 1))
        }
      } catch (e) {
        // 跳过无法访问的文件
      }
    }
  } catch (e) {
    // 跳过无法访问的目录
  }
  return results
}

function buildResult(type, confidence, details, reason) {
  return { type, confidence, details, reason }
}

module.exports = { detectBuildOutput }
