// @edgeone/is-spa/lib/detectors/entries.js
const fs = require('fs')
const path = require('path')
const { detectRouterUsageInFile } = require('./router')
const { BUILD_CONFIG_FILES } = require('../constants')

/**
 * 检测 Vue CLI 项目的入口配置
 * 解析 vue.config.js/ts 中的 pages 字段
 */
function detectVueCliEntries(projectRoot) {
  const configCandidates = BUILD_CONFIG_FILES['vue-cli']
  let configPath = null
  for (const name of configCandidates) {
    const p = path.join(projectRoot, name)
    if (fs.existsSync(p)) {
      configPath = p
      break
    }
  }

  if (!configPath) {
    // 无 vue.config.js/ts → 检查是否是 Vue CLI 项目
    const pkgPath = path.join(projectRoot, 'package.json')
    if (!fs.existsSync(pkgPath)) return null

    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'))
    const allDeps = { ...pkg.dependencies, ...pkg.devDependencies }
    if (!allDeps['@vue/cli-service']) return null

    // 是 Vue CLI 项目但无 pages 配置 → 默认单入口
    const mainPath = path.join(projectRoot, 'src', 'main.js')
    const mainTsPath = path.join(projectRoot, 'src', 'main.ts')
    const entryPath = fs.existsSync(mainPath) ? mainPath : fs.existsSync(mainTsPath) ? mainTsPath : null

    if (!entryPath) return null

    return {
      buildTool: 'vue-cli',
      totalEntryCount: 1,
      entryFiles: [
        {
          name: 'index',
          path: entryPath,
          usesRouter: detectRouterUsageInFile(entryPath),
          outputHtml: 'index.html',
        },
      ],
    }
  }

  // 读取 vue.config.js 内容并提取 pages 配置
  const configContent = fs.readFileSync(configPath, 'utf-8')
  const pages = extractPagesFromVueConfig(configContent, projectRoot)

  if (!pages || Object.keys(pages).length === 0) {
    // 有 vue.config.js 但无 pages → 默认单入口
    const mainPath = path.join(projectRoot, 'src', 'main.js')
    const mainTsPath = path.join(projectRoot, 'src', 'main.ts')
    const entryPath = fs.existsSync(mainPath) ? mainPath : fs.existsSync(mainTsPath) ? mainTsPath : null

    return {
      buildTool: 'vue-cli',
      totalEntryCount: 1,
      entryFiles: entryPath
        ? [
            {
              name: 'index',
              path: entryPath,
              usesRouter: detectRouterUsageInFile(entryPath),
              outputHtml: 'index.html',
            },
          ]
        : [],
    }
  }

  const entryFiles = []
  for (const [name, config] of Object.entries(pages)) {
    const entryStr = typeof config === 'string' ? config : config.entry
    const entryPath = path.isAbsolute(entryStr)
      ? entryStr
      : path.join(projectRoot, entryStr)
    const outputHtml = typeof config === 'string'
      ? `${name}.html`
      : config.filename || `${name}.html`

    entryFiles.push({
      name,
      path: entryPath,
      usesRouter: fs.existsSync(entryPath) ? detectRouterUsageInFile(entryPath) : false,
      outputHtml,
    })
  }

  return {
    buildTool: 'vue-cli',
    totalEntryCount: entryFiles.length,
    entryFiles,
  }
}

/**
 * 从 vue.config.js 内容中提取 pages 配置
 * 策略1：require() 直接加载（准确但需要 node_modules）
 * 策略2：逐行扫描 + 花括号深度计数（不依赖 node_modules）
 */
function extractPagesFromVueConfig(content, projectRoot) {
  // 策略1：尝试使用 require 直接加载配置（仅 .js 文件可 require）
  try {
    const jsConfigPath = path.join(projectRoot, 'vue.config.js')
    if (fs.existsSync(jsConfigPath)) {
      delete require.cache[require.resolve(jsConfigPath)]
      const config = require(jsConfigPath)
      const pages = config.pages || (config.default && config.default.pages)
      if (pages && typeof pages === 'object') {
        return pages
      }
    }
  } catch (e) {
    // require 失败（通常因为 node_modules 未安装），fallback 到文本解析
  }

  // 策略2：逐行扫描 + 花括号深度计数
  // 比正则更可靠，能正确处理嵌套 {} 和多个 page 条目
  const lines = content.split('\n')
  let inPages = false
  let braceDepth = 0
  let currentPage = null
  const pages = {}

  for (const line of lines) {
    const trimmed = line.trim()

    // 跳过注释行
    if (trimmed.startsWith('//') || trimmed.startsWith('*') || trimmed.startsWith('/*')) continue

    // 检测 pages: { 的开始
    if (!inPages) {
      if (/pages\s*:\s*\{/.test(trimmed)) {
        inPages = true
        braceDepth = 1
      }
      continue
    }

    // 统计花括号深度
    for (const ch of trimmed) {
      if (ch === '{') braceDepth++
      if (ch === '}') braceDepth--
    }

    // braceDepth 回到 0 → pages 块结束
    if (braceDepth <= 0) break

    // 深度 2 = page name 层级（pages 的 { 是深度 1，page 的 { 是深度 2）
    // 匹配: index: { 或 404: { 或 'index': {
    const pageNameMatch = trimmed.match(/^['"]?(\w+)['"]?\s*:\s*\{/)
    if (pageNameMatch && braceDepth === 2) {
      currentPage = pageNameMatch[1]
      pages[currentPage] = { entry: null, filename: `${currentPage}.html` }
      continue
    }

    // 深度 2 = page 属性层级
    if (currentPage && braceDepth === 2) {
      const entryMatch = trimmed.match(/entry\s*:\s*['"]([^'"]+)['"]/)
      if (entryMatch) {
        pages[currentPage].entry = entryMatch[1]
      }
      const filenameMatch = trimmed.match(/filename\s*:\s*['"]([^'"]+)['"]/)
      if (filenameMatch) {
        pages[currentPage].filename = filenameMatch[1]
      }
    }

    // 简写形式: index: 'src/main.js'（深度 1，直接在 pages 块下）
    const simpleMatch = trimmed.match(/^['"]?(\w+)['"]?\s*:\s*['"]([^'"]+\.(?:js|ts))['"]/)
    if (simpleMatch && braceDepth === 1) {
      pages[simpleMatch[1]] = {
        entry: simpleMatch[2],
        filename: `${simpleMatch[1]}.html`,
      }
    }
  }

  // 过滤掉没有 entry 的条目
  const validPages = {}
  for (const [name, config] of Object.entries(pages)) {
    if (config.entry) validPages[name] = config
  }

  return Object.keys(validPages).length > 0 ? validPages : null
}
/**
 * 检测 Vite 项目的入口配置
 */
function detectViteEntries(projectRoot) {
  const candidates = BUILD_CONFIG_FILES['vite']
  let configPath = null
  for (const name of candidates) {
    const p = path.join(projectRoot, name)
    if (fs.existsSync(p)) {
      configPath = p
      break
    }
  }

  const pkgPath = path.join(projectRoot, 'package.json')
  if (!fs.existsSync(pkgPath)) return null
  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'))
  const allDeps = { ...pkg.dependencies, ...pkg.devDependencies }
  if (!allDeps['vite']) return null

  // Vite 项目：检查根目录 index.html 和其他 HTML
  const htmlFiles = findHtmlFiles(projectRoot, 1)
  const entryFiles = []

  for (const htmlFile of htmlFiles) {
    const htmlContent = fs.readFileSync(htmlFile, 'utf-8')
    // 从 HTML 中提取 <script type="module" src="...">
    const scriptMatch = htmlContent.match(/<script[^>]*type=["']module["'][^>]*src=["']([^"']+)["']/)
    if (scriptMatch) {
      const entryPath = path.join(projectRoot, scriptMatch[1])
      const name = path.basename(htmlFile, '.html')
      entryFiles.push({
        name: name === 'index' ? 'index' : name,
        path: entryPath,
        usesRouter: fs.existsSync(entryPath) ? detectRouterUsageInFile(entryPath) : false,
        outputHtml: path.relative(projectRoot, htmlFile),
      })
    }
  }

  if (entryFiles.length === 0) return null

  return {
    buildTool: 'vite',
    totalEntryCount: entryFiles.length,
    entryFiles,
  }
}

/**
 * 检测 Webpack 项目的入口配置
 */
function detectWebpackEntries(projectRoot) {
  const configCandidates = BUILD_CONFIG_FILES['webpack']
  let configPath = null
  for (const name of configCandidates) {
    const p = path.join(projectRoot, name)
    if (fs.existsSync(p)) {
      configPath = p
      break
    }
  }
  if (!configPath) return null

  const content = fs.readFileSync(configPath, 'utf-8')

  // 简单匹配 entry 字段
  const entryFiles = []

  // 匹配 entry: { name: 'path' } 形式
  const multiEntryRegex = /entry\s*:\s*\{([^}]+)\}/
  const multiMatch = content.match(multiEntryRegex)
  if (multiMatch) {
    const entriesStr = multiMatch[1]
    const entryRegex = /(\w+)\s*:\s*['"]([^'"]+)['"]/g
    let m
    while ((m = entryRegex.exec(entriesStr)) !== null) {
      const entryPath = path.join(projectRoot, m[2])
      entryFiles.push({
        name: m[1],
        path: entryPath,
        usesRouter: fs.existsSync(entryPath) ? detectRouterUsageInFile(entryPath) : false,
        outputHtml: `${m[1]}.html`,
      })
    }
  }

  // 匹配 entry: 'path' 或 entry: ['path'] 单入口
  if (entryFiles.length === 0) {
    const singleEntryRegex = /entry\s*:\s*['"]([^'"]+)['"]/
    const singleMatch = content.match(singleEntryRegex)
    if (singleMatch) {
      const entryPath = path.join(projectRoot, singleMatch[1])
      entryFiles.push({
        name: 'main',
        path: entryPath,
        usesRouter: fs.existsSync(entryPath) ? detectRouterUsageInFile(entryPath) : false,
        outputHtml: 'index.html',
      })
    }
  }

  if (entryFiles.length === 0) return null

  return {
    buildTool: 'webpack',
    totalEntryCount: entryFiles.length,
    entryFiles,
  }
}

/**
 * 通用入口检测：当无法识别特定构建工具时，扫描常见入口文件
 */
function detectGenericEntries(projectRoot) {
  const candidates = [
    { dir: 'src', files: ['main.js', 'main.ts', 'index.js', 'index.ts', 'index.jsx', 'index.tsx'] },
    { dir: '', files: ['index.js', 'index.ts'] },
  ]

  for (const { dir, files } of candidates) {
    for (const file of files) {
      const filePath = path.join(projectRoot, dir, file)
      if (fs.existsSync(filePath)) {
        return {
          buildTool: 'unknown',
          totalEntryCount: 1,
          entryFiles: [
            {
              name: 'index',
              path: filePath,
              usesRouter: detectRouterUsageInFile(filePath),
              outputHtml: 'index.html',
            },
          ],
        }
      }
    }
  }

  // 扫描 src/pages/ 下的多入口
  const pagesDir = path.join(projectRoot, 'src', 'pages')
  if (fs.existsSync(pagesDir) && fs.statSync(pagesDir).isDirectory()) {
    const entryFiles = []
    const pageDirs = fs.readdirSync(pagesDir).filter((f) => {
      return fs.statSync(path.join(pagesDir, f)).isDirectory()
    })

    for (const pageDir of pageDirs) {
      const mainJs = path.join(pagesDir, pageDir, 'main.js')
      const mainTs = path.join(pagesDir, pageDir, 'main.ts')
      const entryPath = fs.existsSync(mainJs) ? mainJs : fs.existsSync(mainTs) ? mainTs : null
      if (entryPath) {
        entryFiles.push({
          name: pageDir,
          path: entryPath,
          usesRouter: detectRouterUsageInFile(entryPath),
          outputHtml: `${pageDir}.html`,
        })
      }
    }

    if (entryFiles.length > 0) {
      return {
        buildTool: 'unknown',
        totalEntryCount: entryFiles.length,
        entryFiles,
      }
    }
  }

  return null
}

/**
 * 在目录中查找 HTML 文件（非递归或浅递归）
 */
function findHtmlFiles(dir, maxDepth = 0, currentDepth = 0) {
  const results = []
  if (currentDepth > maxDepth) return results
  if (!fs.existsSync(dir)) return results

  const items = fs.readdirSync(dir)
  for (const item of items) {
    if (item === 'node_modules' || item === 'dist' || item === '.git') continue
    const fullPath = path.join(dir, item)
    const stat = fs.statSync(fullPath)
    if (stat.isFile() && item.endsWith('.html')) {
      results.push(fullPath)
    } else if (stat.isDirectory() && currentDepth < maxDepth) {
      results.push(...findHtmlFiles(fullPath, maxDepth, currentDepth + 1))
    }
  }
  return results
}

/**
 * 综合入口检测：按优先级尝试各构建工具
 */
function detectEntries(projectRoot) {
  // 按优先级尝试
  const result =
    detectVueCliEntries(projectRoot) ||
    detectViteEntries(projectRoot) ||
    detectWebpackEntries(projectRoot) ||
    detectGenericEntries(projectRoot)

  if (!result) {
    return {
      buildTool: 'unknown',
      totalEntryCount: 0,
      routerEntryCount: 0,
      entryFiles: [],
    }
  }

  const routerEntryCount = result.entryFiles.filter((e) => e.usesRouter).length

  return {
    ...result,
    routerEntryCount,
  }
}

module.exports = { detectEntries }
