// @edgeone/is-spa/lib/detectors/router.js
const fs = require('fs')
const path = require('path')
const { ROUTER_LIBRARIES, ROUTER_IMPORT_PATTERNS } = require('../constants')

/**
 * 检测 package.json 中的路由库依赖
 */
function detectRouterDependency(projectRoot) {
  const pkgPath = path.join(projectRoot, 'package.json')
  if (!fs.existsSync(pkgPath)) {
    return { hasRouterDependency: false, routerLibrary: null }
  }

  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'))
  const allDeps = {
    ...pkg.dependencies,
    ...pkg.devDependencies,
  }

  for (const lib of ROUTER_LIBRARIES) {
    if (allDeps[lib]) {
      return { hasRouterDependency: true, routerLibrary: lib }
    }
  }

  return { hasRouterDependency: false, routerLibrary: null }
}

/**
 * 检测 src/router/ 目录是否存在
 */
function detectRouterDirectory(projectRoot) {
  const candidates = [
    path.join(projectRoot, 'src', 'router'),
    path.join(projectRoot, 'src', 'routes'),
    path.join(projectRoot, 'app', 'router'),
    path.join(projectRoot, 'app', 'routes'),
  ]

  for (const dir of candidates) {
    if (fs.existsSync(dir) && fs.statSync(dir).isDirectory()) {
      return { hasRouterDirectory: true, routerDirectoryPath: dir }
    }
  }

  return { hasRouterDirectory: false, routerDirectoryPath: null }
}

/**
 * 检测入口文件中是否引入了 router
 */
function detectRouterUsageInFile(filePath) {
  if (!fs.existsSync(filePath)) {
    return false
  }

  const content = fs.readFileSync(filePath, 'utf-8')
  return ROUTER_IMPORT_PATTERNS.some((pattern) => pattern.test(content))
}

/**
 * 综合路由检测
 */
function detectRouter(projectRoot) {
  const dep = detectRouterDependency(projectRoot)
  const dir = detectRouterDirectory(projectRoot)

  return {
    hasRouterDependency: dep.hasRouterDependency,
    routerLibrary: dep.routerLibrary,
    hasRouterDirectory: dir.hasRouterDirectory,
    routerDirectoryPath: dir.routerDirectoryPath,
    // hasRouterUsageInEntry 由 entry detector 填充
    hasRouterUsageInEntry: false,
  }
}

module.exports = {
  detectRouter,
  detectRouterDependency,
  detectRouterDirectory,
  detectRouterUsageInFile,
}
