// @edgeone/pages-agent-toolkit/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  SandboxNotAvailableError: () => SandboxNotAvailableError,
  buildSandboxProxy: () => buildSandboxProxy,
  buildTools: () => buildTools,
  isToolErrorResult: () => isToolErrorResult,
  toToolErrorResult: () => toToolErrorResult
});
module.exports = __toCommonJS(index_exports);

// src/sandbox/types.ts
var SandboxNotAvailableError = class _SandboxNotAvailableError extends Error {
  constructor(message, options = {}) {
    super(_SandboxNotAvailableError.composeMessage(message, options));
    this.name = "SandboxNotAvailableError";
    this.code = options.code || "SANDBOX_UNKNOWN_ERROR";
    this.operation = options.operation;
    this.status = options.status;
    this.businessCode = options.businessCode;
    this.businessMessage = options.businessMessage;
    this.requestId = options.requestId;
    this.details = options.details;
    this.cause = options.cause;
  }
  /**
   * 在原始报错文本后拼接结构化诊断字段，使 message 自包含。
   *
   * 这样即使调用方用 `String(error)`、打印 `error.message` 或在旧环境里以
   * 有损方式序列化，也能看到 code/operation/status/businessCode/requestId 等关键信息。
   */
  static composeMessage(message, options) {
    const base = typeof message === "string" ? message : String(message ?? "");
    const parts = [];
    const code = options.code || "SANDBOX_UNKNOWN_ERROR";
    parts.push(`code=${code}`);
    if (options.operation) parts.push(`operation=${options.operation}`);
    if (options.status !== void 0) parts.push(`status=${options.status}`);
    if (options.businessCode !== void 0) parts.push(`businessCode=${options.businessCode}`);
    if (options.requestId) parts.push(`requestId=${options.requestId}`);
    return parts.length ? `${base} (${parts.join(", ")})` : base;
  }
  /**
   * 序列化为包含全部诊断字段的纯对象。
   *
   * `JSON.stringify(error)` 会自动调用此方法，从而避免默认序列化丢失
   * 不可枚举的 message 以及结构化字段的问题。
   */
  toJSON() {
    const result = {
      name: this.name,
      message: this.message,
      code: this.code
    };
    if (this.operation !== void 0) result.operation = this.operation;
    if (this.status !== void 0) result.status = this.status;
    if (this.businessCode !== void 0) result.businessCode = this.businessCode;
    if (this.businessMessage !== void 0) result.businessMessage = this.businessMessage;
    if (this.requestId !== void 0) result.requestId = this.requestId;
    if (this.details !== void 0) result.details = this.details;
    return result;
  }
};

// src/sandbox/api-domain.ts
var SANDBOX_API_BASES = {
  test: {
    china: "https://eo-test.qcloud.com/v1/sandbox",
    global: "https://test-api.edgeone.ai/v1/sandbox"
  },
  production: {
    china: "https://pages-api.cloud.tencent.com/v1/sandbox",
    global: "https://pages-api.edgeone.ai/v1/sandbox"
  },
  pre: {
    china: "https://pages-api-pre.cloud.tencent.com/v1/sandbox",
    global: "https://pre-api.edgeone.ai/v1/sandbox"
  }
};
var SITE_ALIASES = {
  china: "china",
  domestic: "china",
  cn: "china",
  global: "global",
  intl: "global",
  overseas: "global"
};
function resolveSandboxApiBase(config) {
  const explicitBase = firstPresent(config.apiBase, getEnv("SANDBOX_API_BASE"));
  if (explicitBase) return normalizeSandboxApiBase(explicitBase);
  const site = resolveSandboxSite(config);
  const apiEnv = resolveSandboxApiEnv(config);
  return SANDBOX_API_BASES[apiEnv][site];
}
function resolveDefaultSandboxRegion(config) {
  if (config.region) return config.region;
  return resolveSandboxSite(config) === "global" ? "ap-singapore" : "ap-beijing";
}
function normalizeSandboxApiBase(apiBase) {
  const trimmed = apiBase.trim().replace(/\/+$/, "");
  if (!trimmed) return "";
  if (!/^https?:\/\//i.test(trimmed)) {
    throw new SandboxNotAvailableError(
      `Invalid SANDBOX_API_BASE: ${apiBase}. Expected an http(s) URL.`,
      {
        code: "SANDBOX_INVALID_PARAMETER",
        operation: "resolve_api_base",
        details: { apiBasePresent: true }
      }
    );
  }
  if (trimmed.endsWith("/v1/sandbox")) return trimmed;
  if (trimmed.endsWith("/v1")) return `${trimmed}/sandbox`;
  return `${trimmed}/v1/sandbox`;
}
function resolveSandboxSite(config) {
  const site = firstPresent(
    config.site,
    getEnv("SANDBOX_SITE"),
    getEnv("EDGEONE_PAGES_API_REGION")
  );
  if (!site) return "china";
  const normalized = SITE_ALIASES[site.toLowerCase()];
  if (!normalized) {
    throw new SandboxNotAvailableError(`Invalid sandbox site: ${site}. Expected china/global.`, {
      code: "SANDBOX_INVALID_PARAMETER",
      operation: "resolve_api_base",
      details: { site }
    });
  }
  return normalized;
}
function resolveSandboxApiEnv(config) {
  const explicitEnv = firstPresent(getEnv("SANDBOX_API_ENV"), config.apiEnv, getEnv("API_ENV")).toLowerCase();
  if (explicitEnv) {
    if (explicitEnv === "test" || explicitEnv === "production" || explicitEnv === "pre") return explicitEnv;
    throw new SandboxNotAvailableError(`Invalid sandbox apiEnv: ${explicitEnv}. Expected test/production/pre.`, {
      code: "SANDBOX_INVALID_PARAMETER",
      operation: "resolve_api_base",
      details: { apiEnv: explicitEnv }
    });
  }
  return config.authToken.trim().startsWith("sandbox.v1.") ? "production" : "test";
}
function firstPresent(...values) {
  for (const value of values) {
    if (typeof value !== "string") continue;
    const trimmed = value.trim();
    if (trimmed) return trimmed;
  }
  return "";
}
function getEnv(key) {
  return globalThis.process?.env?.[key] || "";
}

// src/sandbox/pages-api-client.ts
var DEFAULT_TIMEOUT = 300;
var SEALED_TOKEN_PREFIX = "sandbox.v1.";
var PagesApiClient = class {
  constructor(config) {
    const authToken = String(config.authToken || "").trim();
    if (!authToken) {
      throw new SandboxNotAvailableError("sandbox auth token missing", {
        code: "SANDBOX_INVALID_PARAMETER",
        operation: "init"
      });
    }
    this.config = { ...config, authToken };
    try {
      this.apiBase = resolveSandboxApiBase(this.config);
    } catch (error) {
      if (error instanceof SandboxNotAvailableError) throw error;
      throw new SandboxNotAvailableError(formatError(error), {
        code: "SANDBOX_INVALID_PARAMETER",
        operation: "resolve_api_base",
        cause: error
      });
    }
    this.headers = {
      Authorization: `Bearer ${authToken}`,
      "Content-Type": "application/json"
    };
  }
  /**
   * 申请沙箱实例。
   *
   * 调用后端 acquire 接口获取可用沙箱，并返回实例 ID、访问 Token、过期时间和沙箱域名等信息。
   */
  async acquire() {
    if (!this.apiBase) {
      throw new SandboxNotAvailableError(
        "SANDBOX_API_BASE is not configured; unable to acquire a sandbox instance",
        {
          code: "SANDBOX_INVALID_PARAMETER",
          operation: "acquire"
        }
      );
    }
    const payload = this.buildAcquirePayload();
    const resp = await fetch(this.buildUrl("/acquire"), {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(payload)
    }).catch((error) => {
      throw new SandboxNotAvailableError(`Failed to acquire sandbox instance: ${formatError(error)}`, {
        code: "SANDBOX_NETWORK_ERROR",
        operation: "acquire",
        cause: error
      });
    });
    if (!resp.ok) {
      const errorBody = await readErrorBody(resp);
      throw new SandboxNotAvailableError(errorBody.message, {
        code: "SANDBOX_NETWORK_ERROR",
        operation: "acquire",
        status: resp.status,
        businessCode: errorBody.businessCode,
        businessMessage: errorBody.message,
        requestId: errorBody.requestId
      });
    }
    const raw = await resp.json().catch(() => null);
    const data = extractEnvelopeData(raw, "acquire");
    if (!data) {
      throw new SandboxNotAvailableError("Invalid acquire response from sandbox API", {
        code: "SANDBOX_INTERNAL_ERROR",
        operation: "acquire",
        requestId: extractEnvelopeRequestId(raw)
      });
    }
    if (isBusinessError(data)) {
      const message = data.message || "Unknown sandbox API error";
      throw new SandboxNotAvailableError(message, {
        code: mapBackendCode(data.code),
        operation: "acquire",
        businessCode: data.code,
        businessMessage: message,
        requestId: extractEnvelopeRequestId(raw)
      });
    }
    assertAcquireResponse(data);
    return {
      instanceId: data.InstanceId,
      sandboxToken: data.Token,
      expiresAt: data.InstanceExpiresAt,
      ...data.TrafficToken && { trafficToken: data.TrafficToken },
      ...data.SandboxDomain && { sandboxDomain: data.SandboxDomain },
      ...!data.SandboxDomain && this.config.sandboxDomain && { sandboxDomain: this.config.sandboxDomain },
      ...data.EnvdVersion && { envdVersion: data.EnvdVersion },
      ...!data.EnvdVersion && this.config.envdVersion && { envdVersion: this.config.envdVersion }
    };
  }
  /**
   * 释放沙箱实例。
   *
   * 调用后端 release 接口释放指定沙箱；释放失败时返回 success=false，不向上抛出异常。
   */
  async release(instanceId) {
    if (!this.apiBase) return { success: false };
    const resp = await fetch(this.buildUrl("/release"), {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({ InstanceId: instanceId })
    }).catch(() => null);
    if (!resp || !resp.ok) return { success: false };
    let data = null;
    try {
      data = extractEnvelopeData(await resp.json().catch(() => null), "release");
    } catch {
      return { success: false };
    }
    if (!data) return { success: false };
    if (isBusinessError(data)) return { success: false };
    return { success: Boolean(data.Success) };
  }
  /**
   * 续期沙箱实例。
   *
   * 调用后端 update 接口为指定沙箱延长有效期，并返回后端确认的新过期时间。
   */
  async update(instanceId, seconds) {
    if (!this.apiBase) {
      throw new SandboxNotAvailableError(
        "SANDBOX_API_BASE is not configured; unable to renew the sandbox instance",
        {
          code: "SANDBOX_INVALID_PARAMETER",
          operation: "update"
        }
      );
    }
    const timeout = normalizeTimeoutSeconds(seconds, "update");
    const resp = await fetch(this.buildUrl("/update"), {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({ InstanceId: instanceId, Timeout: timeout })
    }).catch((error) => {
      throw new SandboxNotAvailableError(`Failed to renew sandbox instance: ${formatError(error)}`, {
        code: "SANDBOX_NETWORK_ERROR",
        operation: "update",
        cause: error
      });
    });
    if (!resp.ok) {
      const errorBody = await readErrorBody(resp);
      throw new SandboxNotAvailableError(errorBody.message, {
        code: "SANDBOX_NETWORK_ERROR",
        operation: "update",
        status: resp.status,
        businessMessage: errorBody.message,
        requestId: errorBody.requestId
      });
    }
    const raw = await resp.json().catch(() => null);
    const data = extractEnvelopeData(raw, "update");
    if (!data) {
      throw new SandboxNotAvailableError("Invalid update response from sandbox API", {
        code: "SANDBOX_INTERNAL_ERROR",
        operation: "update",
        requestId: extractEnvelopeRequestId(raw)
      });
    }
    if (isBusinessError(data)) {
      const message = data.message || "Unknown sandbox API error";
      throw new SandboxNotAvailableError(message, {
        code: mapBackendCode(data.code),
        operation: "update",
        businessCode: data.code,
        businessMessage: message,
        requestId: extractEnvelopeRequestId(raw)
      });
    }
    assertUpdateResponse(data);
    return {
      instanceId: data.InstanceId,
      expiresAt: data.InstanceExpiresAt,
      ...data.Message && { message: data.Message }
    };
  }
  buildUrl(path) {
    return `${this.apiBase}${path}`;
  }
  buildAcquirePayload() {
    const conversationId = normalizeRequiredString(this.config.conversationId, "ConversationId");
    const projectId = firstPresentString(
      this.config.projectId,
      getEnv2("PROJECT_ID")
    );
    const deploymentId = normalizeOptionalString(this.config.deploymentId);
    const payload = {
      ProjectId: projectId,
      DeploymentId: deploymentId,
      ConversationId: conversationId,
      Region: resolveDefaultSandboxRegion(this.config),
      Timeout: normalizeTimeoutSeconds(this.config.timeout ?? DEFAULT_TIMEOUT, "acquire")
    };
    if (requiresProjectId(this.config.authToken) && !payload.ProjectId) {
      throw new SandboxNotAvailableError(
        "ProjectId is required. Set PROJECT_ID or pass projectId when initializing the SDK.",
        {
          code: "SANDBOX_INVALID_PARAMETER",
          operation: "acquire",
          businessMessage: "ProjectId is required"
        }
      );
    }
    return compactPayload(payload);
  }
};
async function readErrorBody(resp) {
  const fallback = `Sandbox API request failed with status ${resp.status}`;
  const headerRequestId = extractRequestIdFromHeaders(resp);
  const text = await resp.text().catch(() => "");
  if (!text.trim()) return { message: resp.statusText || fallback, requestId: headerRequestId };
  try {
    const body = JSON.parse(text);
    const message = firstPresentString(
      body.error,
      body.Data?.message,
      body.data?.message,
      body.message,
      text
    ) || fallback;
    const requestId = firstPresentString(
      headerRequestId,
      body.requestId,
      body.request_id,
      body.RequestId,
      body.Data?.requestId,
      body.Data?.request_id,
      body.data?.requestId,
      body.data?.request_id
    );
    const businessCode = typeof body.Code === "number" ? body.Code : typeof body.code === "number" ? body.code : void 0;
    return { message, requestId, businessCode };
  } catch {
    return { message: text || resp.statusText || fallback, requestId: headerRequestId };
  }
}
function extractRequestIdFromHeaders(resp) {
  const headers = resp.headers;
  if (!headers || typeof headers.get !== "function") return void 0;
  return firstPresentString(
    headers.get("x-request-id"),
    headers.get("x-trace-id"),
    headers.get("request-id")
  );
}
function extractEnvelopeRequestId(raw) {
  if (!raw || typeof raw !== "object") return void 0;
  const envelope = raw;
  return firstPresentString(envelope.requestId, envelope.request_id, envelope.RequestId);
}
function formatError(error) {
  return error instanceof Error ? error.message : String(error);
}
function extractEnvelopeData(raw, operation) {
  if (!raw || typeof raw !== "object") return null;
  if (isGatewayError(raw)) {
    const message = raw.message || "Unknown sandbox API error";
    throw new SandboxNotAvailableError(message, {
      code: "SANDBOX_INTERNAL_ERROR",
      operation,
      businessCode: raw.Code,
      businessMessage: message,
      requestId: extractEnvelopeRequestId(raw)
    });
  }
  if ("Data" in raw) {
    const data = raw.Data;
    return data && typeof data === "object" ? data : null;
  }
  if ("data" in raw) {
    const data = raw.data;
    return data && typeof data === "object" ? data : null;
  }
  return raw;
}
function isGatewayError(value) {
  return Boolean(
    value && typeof value === "object" && "Code" in value && typeof value.Code === "number" && value.Code !== 0
  );
}
function isBusinessError(value) {
  if (!value || typeof value !== "object") return false;
  const code = value.code;
  return typeof code === "string" && code.trim() !== "";
}
function normalizeTimeoutSeconds(value, operation) {
  if (typeof value !== "number" || !Number.isFinite(value) || !Number.isInteger(value) || value <= 0) {
    throw new SandboxNotAvailableError(`${operation} Timeout must be a positive integer number of seconds`, {
      code: "SANDBOX_INVALID_PARAMETER",
      operation
    });
  }
  return value;
}
function normalizeRequiredString(value, fieldName) {
  const normalized = normalizeOptionalString(value);
  if (!normalized) {
    throw new SandboxNotAvailableError(`Missing ${fieldName}`, {
      code: "SANDBOX_INVALID_PARAMETER",
      operation: "acquire"
    });
  }
  return normalized;
}
function firstPresentString(...values) {
  for (const value of values) {
    const normalized = normalizeOptionalString(value);
    if (normalized) return normalized;
  }
  return void 0;
}
function normalizeOptionalString(value) {
  if (typeof value !== "string") return void 0;
  const trimmed = value.trim();
  return trimmed || void 0;
}
function requiresProjectId(authToken) {
  return !authToken.trim().startsWith(SEALED_TOKEN_PREFIX);
}
function compactPayload(payload) {
  return Object.fromEntries(
    Object.entries(payload).filter(([, value]) => value !== void 0)
  );
}
function getEnv2(key) {
  return globalThis.process?.env?.[key];
}
function assertAcquireResponse(data) {
  if (!data.InstanceId || !data.Token || !data.InstanceExpiresAt) {
    throw new SandboxNotAvailableError("Invalid acquire response from sandbox API", {
      code: "SANDBOX_INTERNAL_ERROR",
      operation: "acquire"
    });
  }
}
function assertUpdateResponse(data) {
  if (!data.InstanceId || !data.InstanceExpiresAt) {
    throw new SandboxNotAvailableError("Invalid update response from sandbox API", {
      code: "SANDBOX_INTERNAL_ERROR",
      operation: "update"
    });
  }
}
var KNOWN_BACKEND_CODES = /* @__PURE__ */ new Set([
  "SANDBOX_INVALID_PARAMETER",
  "SANDBOX_AUTHORIZATION_EXPIRED",
  "SANDBOX_RESOURCE_NOT_FOUND",
  "SANDBOX_LIMIT_EXCEEDED",
  "SANDBOX_FAILED_OPERATION",
  "SANDBOX_INTERNAL_ERROR"
]);
function mapBackendCode(rawCode) {
  const raw = String(rawCode ?? "").trim();
  if (!raw) return "SANDBOX_UNKNOWN_ERROR";
  const base = raw.split(".")[0];
  const snake = base.replace(/([a-z0-9])([A-Z])/g, "$1_$2").toUpperCase();
  const candidate = `SANDBOX_${snake}`;
  return KNOWN_BACKEND_CODES.has(candidate) ? candidate : "SANDBOX_UNKNOWN_ERROR";
}

// src/sandbox/instance-cache.ts
var InstanceCache = class {
  constructor() {
    this.cache = /* @__PURE__ */ new Map();
  }
  get(conversationId) {
    const entry = this.cache.get(conversationId);
    if (!entry) return null;
    if (entry.expiresAt.getTime() <= Date.now()) {
      this.cache.delete(conversationId);
      return null;
    }
    return entry;
  }
  set(conversationId, entry) {
    this.cache.set(conversationId, entry);
  }
  delete(conversationId) {
    this.cache.delete(conversationId);
  }
  updateExpiresAt(conversationId, expiresAt) {
    const entry = this.cache.get(conversationId);
    if (!entry) return;
    entry.expiresAt = new Date(expiresAt);
    entry.info = { ...entry.info, expiresAt };
  }
  getInfo(conversationId) {
    const entry = this.get(conversationId);
    return entry ? { ...entry.info } : null;
  }
};

// src/sandbox/e2b-factory.ts
var BROWSER_RESULT_PREFIX = "__PAGES_BROWSER_RESULT__=";
async function importRuntimeModule(specifier) {
  return await import(specifier);
}
async function createSandboxClient(info) {
  let Sandbox;
  try {
    ({ Sandbox } = await importRuntimeModule("@e2b/code-interpreter"));
  } catch (error) {
    throw new SandboxNotAvailableError(`Failed to load sandbox runtime client: ${formatError2(error)}`, {
      code: "SANDBOX_CLIENT_ERROR",
      operation: "create_client",
      details: summarizeAcquireInfo(info),
      cause: error
    });
  }
  let sandbox;
  try {
    sandbox = new Sandbox({
      sandboxId: info.instanceId,
      sandboxDomain: info.sandboxDomain || "ags.tencentcloudapi.com",
      envdAccessToken: info.sandboxToken,
      trafficAccessToken: info.trafficToken || void 0,
      envdVersion: info.envdVersion || "0.2.4"
    });
  } catch (error) {
    throw wrapSandboxClientError(error, "create_client", "Failed to create sandbox runtime client", info);
  }
  const envdToken = sandbox.envdAccessToken || sandbox._envdAccessToken || info.sandboxToken || "";
  const getHost = (port) => {
    try {
      return sandbox.getHost(port);
    } catch (error) {
      throw wrapSandboxClientError(error, "create_client", `Failed to resolve sandbox host for port ${port}`, info);
    }
  };
  const browserHost = getHost(9e3);
  const cdpUrl = `https://${browserHost}/cdp?access_token=${envdToken}`;
  const liveUrl = `https://${browserHost}/novnc/vnc_lite.html?access_token=${envdToken}&path=websockify%3Faccess_token%3D${envdToken}`;
  async function runBrowserOp(op, payload = {}) {
    try {
      const encoded = globalThis.Buffer.from(JSON.stringify({ op, token: envdToken, ...payload }), "utf-8").toString("base64");
      const command = BROWSER_RUNNER_SCRIPT.replace("__PAYLOAD_B64__", encoded);
      const result = await sandbox.commands.run(command, { timeoutMs: 6e4 });
      const stdout = String(result?.stdout || "");
      for (const line of stdout.split(/\r?\n/).reverse()) {
        if (!line.startsWith(BROWSER_RESULT_PREFIX)) continue;
        const data = JSON.parse(line.slice(BROWSER_RESULT_PREFIX.length));
        if (!data.ok) throw new Error(data.error || "browser operation failed");
        return data.result || {};
      }
      throw new Error(
        `browser operation returned no result. stdout=${stdout.slice(0, 300)} stderr=${String(result?.stderr || "").slice(0, 300)}`
      );
    } catch (error) {
      throw wrapSandboxClientError(error, "browser", `Sandbox browser operation failed: ${op}`, info);
    }
  }
  return {
    commands: {
      async run(command, opts) {
        try {
          const result = await sandbox.commands.run(command, {
            cwd: opts?.cwd,
            envs: opts?.env,
            user: opts?.user,
            timeoutMs: opts?.timeout ? opts.timeout * 1e3 : void 0,
            onStdout: opts?.onStdout ? (msg) => opts.onStdout(msg.line || msg) : void 0,
            onStderr: opts?.onStderr ? (msg) => opts.onStderr(msg.line || msg) : void 0
          });
          return {
            stdout: result.stdout,
            stderr: result.stderr,
            exitCode: result.exitCode
          };
        } catch (error) {
          throw wrapSandboxClientError(error, "command", "Sandbox command failed", info);
        }
      }
    },
    files: {
      async read(path) {
        try {
          return await sandbox.files.read(path);
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox file read failed: ${path}`, info);
        }
      },
      async write(path, content) {
        try {
          await sandbox.files.write(path, content);
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox file write failed: ${path}`, info);
        }
      },
      async list(path) {
        try {
          const entries = await sandbox.files.list(path);
          return entries.map((e) => ({
            name: e.name,
            type: e.type,
            path: e.path
          }));
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox file list failed: ${path}`, info);
        }
      },
      async makeDir(path) {
        try {
          await sandbox.files.makeDir(path);
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox directory create failed: ${path}`, info);
        }
      },
      async exists(path) {
        try {
          return await sandbox.files.exists(path);
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox file exists failed: ${path}`, info);
        }
      },
      async remove(path) {
        try {
          await sandbox.files.remove(path);
        } catch (error) {
          throw wrapSandboxClientError(error, "file", `Sandbox file remove failed: ${path}`, info);
        }
      }
    },
    browser: {
      get cdpUrl() {
        return cdpUrl;
      },
      get liveUrl() {
        return liveUrl;
      },
      async goto(url, opts) {
        const data = await runBrowserOp("goto", {
          url,
          waitUntil: opts?.waitUntil || "domcontentloaded"
        });
        return { success: true, data };
      },
      async screenshot(opts) {
        const data = await runBrowserOp("screenshot", { fullPage: Boolean(opts?.fullPage) });
        return { base64Image: String(data.base64Image || "") };
      },
      async click(selector) {
        await runBrowserOp("click", { selector });
        return { success: true };
      },
      async type(selector, text) {
        await runBrowserOp("type", { selector, text });
        return { success: true };
      },
      async evaluate(script) {
        const data = await runBrowserOp("evaluate", { script });
        return { success: true, data: data.data };
      },
      async getContent() {
        const data = await runBrowserOp("getContent");
        return { success: true, data: data.data || "" };
      },
      async close() {
        await runBrowserOp("close");
      }
    },
    async runCode(code, opts) {
      try {
        const execution = await sandbox.runCode(code, {
          language: opts?.language || "python",
          timeoutMs: opts?.timeout ? opts.timeout * 1e3 : void 0,
          onStdout: opts?.onStdout ? (msg) => opts.onStdout(msg.line || msg) : void 0,
          onStderr: opts?.onStderr ? (msg) => opts.onStderr(msg.line || msg) : void 0
        });
        return {
          results: execution.results.map((r) => r.toJSON ? r.toJSON() : r),
          logs: {
            stdout: execution.logs.stdout.map((l) => l.line || l),
            stderr: execution.logs.stderr.map((l) => l.line || l)
          },
          error: execution.error ? {
            name: execution.error.name,
            message: execution.error.value,
            traceback: execution.error.traceback
          } : void 0
        };
      } catch (error) {
        throw wrapSandboxClientError(error, "run_code", "Sandbox runCode failed", info);
      }
    },
    getHost,
    get envdAccessToken() {
      return envdToken;
    },
    async kill() {
      try {
        await sandbox.kill();
      } catch (error) {
        throw wrapSandboxClientError(error, "release", "Sandbox kill failed", info);
      }
    }
  };
}
var BROWSER_RUNNER_SCRIPT = String.raw`python3 - <<'PY'
import asyncio
import base64
import json

PREFIX = "__PAGES_BROWSER_RESULT__="

async def main():
    payload = json.loads(base64.b64decode("__PAYLOAD_B64__").decode("utf-8"))
    try:
        from playwright.async_api import async_playwright
    except Exception as exc:
        raise RuntimeError(
            "playwright is not installed inside sandbox; browser tool requires sandbox image with Playwright support"
        ) from exc

    token = payload.get("token") or ""
    cdp_url = f"http://127.0.0.1:9000/cdp?access_token={token}"
    playwright = await async_playwright().start()
    try:
        browser = await playwright.chromium.connect_over_cdp(
            cdp_url,
            headers={"X-Access-Token": token} if token else None,
        )
        contexts = browser.contexts
        if contexts:
            context = contexts[0]
            page = context.pages[0] if context.pages else await context.new_page()
        else:
            context = await browser.new_context()
            page = await context.new_page()

        op = payload.get("op")
        result = {}
        if op == "goto":
            response = await page.goto(payload["url"], wait_until=payload.get("waitUntil") or "domcontentloaded")
            result = {"url": page.url, "title": await page.title(), "status": response.status if response else None}
        elif op == "screenshot":
            buf = await page.screenshot(full_page=bool(payload.get("fullPage")))
            result = {"base64Image": base64.b64encode(buf).decode("ascii")}
        elif op == "click":
            await page.click(payload["selector"])
            result = {"success": True}
        elif op == "type":
            await page.fill(payload["selector"], payload.get("text") or "")
            result = {"success": True}
        elif op == "evaluate":
            result = {"data": await page.evaluate(payload["script"])}
        elif op == "getContent":
            result = {"data": await page.content()}
        elif op == "close":
            await page.close()
            result = {"success": True}
        else:
            raise ValueError(f"unsupported browser op: {op}")
        print(PREFIX + json.dumps({"ok": True, "result": result}, ensure_ascii=False), flush=True)
    except Exception as exc:
        print(PREFIX + json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False), flush=True)
        raise
    finally:
        try:
            await playwright.stop()
        except Exception:
            pass

asyncio.run(main())
PY`;
function wrapSandboxClientError(error, operation, message, info) {
  if (error instanceof SandboxNotAvailableError) return error;
  const formatted = formatError2(error);
  const code = classifyClientError(error, formatted, operation);
  return new SandboxNotAvailableError(`${message}${formatSandboxContext(info)}: ${formatted}`, {
    code,
    operation,
    details: summarizeAcquireInfo(info),
    cause: error
  });
}
function formatSandboxContext(info) {
  const parts = [];
  if (info?.instanceId) parts.push(`instanceId=${info.instanceId}`);
  if (info?.expiresAt) parts.push(`expiresAt=${info.expiresAt}`);
  return parts.length ? ` [${parts.join(", ")}]` : "";
}
function classifyClientError(error, message, operation) {
  const lower = message.toLowerCase();
  const code = String(error?.code || "").toLowerCase();
  const status = Number(error?.status ?? error?.statusCode ?? error?.response?.status);
  if (lower.includes("token expired") || lower.includes("invalid sandbox token") || lower.includes("invalid token") || lower.includes("instance not found") || lower.includes("sandbox not found") || lower.includes("http 404") || lower.includes("not authorized") || lower.includes("unauthorized") || lower.includes("forbidden") || lower.includes("expired") || status === 404 || status === 401 || status === 403) return "SANDBOX_INSTANCE_UNAVAILABLE";
  if (code === "econnrefused" || code === "econnreset" || code === "enotfound" || code === "etimedout" || lower.includes("network") || lower.includes("timeout") || lower.includes("connection")) return "SANDBOX_NETWORK_ERROR";
  return operation === "create_client" ? "SANDBOX_CLIENT_ERROR" : "SANDBOX_UNKNOWN_ERROR";
}
function summarizeAcquireInfo(info) {
  return Object.fromEntries(Object.entries({
    instanceId: info.instanceId,
    sandboxId: info.instanceId,
    sandboxDomain: info.sandboxDomain,
    envdVersion: info.envdVersion,
    expiresAt: info.expiresAt,
    hasSandboxToken: Boolean(info.sandboxToken),
    hasTrafficToken: Boolean(info.trafficToken)
  }).filter(([, value]) => value !== void 0 && value !== ""));
}
function formatError2(error) {
  return error instanceof Error ? error.message : String(error);
}

// src/sandbox/index.ts
function isSandboxDebugEnabled() {
  const env = globalThis.process?.env?.MAKERS_AGENT_TOOLKIT_DEBUG;
  return env === "1" || env === "true";
}
function summarizeSandboxInstance(info, client) {
  return Object.fromEntries(Object.entries({
    instanceId: info.instanceId,
    sandboxDomain: info.sandboxDomain,
    envdVersion: info.envdVersion,
    expiresAt: info.expiresAt,
    clientType: client?.constructor?.name
  }).filter(([, value]) => value !== void 0 && value !== ""));
}
function sandboxDebugLog(event, data) {
  if (!isSandboxDebugEnabled()) return;
  console.error(`[pages-agent-toolkit] ${event} ${JSON.stringify(data)}`);
}
function summarizeSandboxError(error) {
  return Object.fromEntries(Object.entries({
    name: error?.name,
    code: error?.code,
    operation: error?.operation,
    status: error?.status,
    businessCode: error?.businessCode,
    businessMessage: error?.businessMessage,
    requestId: error?.requestId,
    instanceId: error?.details?.instanceId,
    expiresAt: error?.details?.expiresAt,
    message: error?.message
  }).filter(([, value]) => value !== void 0 && value !== ""));
}
var LazySandbox = class {
  constructor(config) {
    this._instance = null;
    this._initializing = null;
    this.config = config;
    this.apiClient = new PagesApiClient(config);
    this.cache = new InstanceCache();
  }
  async getClient() {
    const cached = this.cache.get(this.config.conversationId);
    if (cached) {
      sandboxDebugLog("sandbox.cache.hit", { conversationIdPresent: Boolean(this.config.conversationId), instanceId: cached.instanceId, expiresAt: cached.expiresAt.toISOString() });
      this._instance = cached.sandbox;
      return cached.sandbox;
    }
    sandboxDebugLog("sandbox.cache.miss", { conversationIdPresent: Boolean(this.config.conversationId) });
    if (this._initializing) {
      sandboxDebugLog("sandbox.acquire.join", { conversationIdPresent: Boolean(this.config.conversationId) });
      return this._initializing;
    }
    this._initializing = this._acquire();
    try {
      return await this._initializing;
    } finally {
      this._initializing = null;
    }
  }
  async _acquire() {
    sandboxDebugLog("sandbox.acquire.start", { conversationIdPresent: Boolean(this.config.conversationId) });
    let info;
    let client;
    try {
      info = await this.apiClient.acquire();
      client = await createSandboxClient(info);
    } catch (error) {
      sandboxDebugLog("sandbox.acquire.failure", { error: summarizeSandboxError(error), conversationIdPresent: Boolean(this.config.conversationId) });
      throw error;
    }
    const entry = {
      instanceId: info.instanceId,
      sandboxToken: info.sandboxToken,
      trafficToken: info.trafficToken,
      sandboxDomain: info.sandboxDomain,
      envdVersion: info.envdVersion,
      expiresAt: new Date(info.expiresAt),
      sandbox: client,
      info: { ...info }
    };
    this.cache.set(this.config.conversationId, entry);
    this._instance = client;
    sandboxDebugLog("sandbox.acquire.success", { sandbox: summarizeSandboxInstance(info, client) });
    return client;
  }
  async withClient(operation) {
    const client = await this.getClient();
    try {
      return await operation(client);
    } catch (error) {
      if (!isRecoverableSandboxError(error)) throw error;
      sandboxDebugLog("sandbox.cache.invalidate", { reason: "data_plane_error", error: summarizeSandboxError(error), conversationIdPresent: Boolean(this.config.conversationId) });
      this.cache.delete(this.config.conversationId);
      this._instance = null;
      sandboxDebugLog("sandbox.acquire.retry", { reason: "data_plane_error", error: summarizeSandboxError(error), conversationIdPresent: Boolean(this.config.conversationId) });
      const refreshedClient = await this.getClient();
      try {
        return await operation(refreshedClient);
      } catch (retryError) {
        if (isRecoverableSandboxError(retryError)) {
          sandboxDebugLog("sandbox.cache.invalidate", { reason: "data_plane_retry_error", error: summarizeSandboxError(retryError), conversationIdPresent: Boolean(this.config.conversationId) });
          this.cache.delete(this.config.conversationId);
          this._instance = null;
        }
        throw retryError;
      }
    }
  }
  getInfo() {
    const info = this.cache.getInfo(this.config.conversationId);
    if (!info) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
    return info;
  }
  async extendTimeout(seconds) {
    validateExtendSeconds(seconds);
    const client = await this.getClient();
    const cached = this.cache.get(this.config.conversationId);
    if (!cached) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
    try {
      sandboxDebugLog("sandbox.renew.start", { instanceId: cached.instanceId, seconds });
      const renewed = await this.apiClient.update(cached.instanceId, seconds);
      this.cache.updateExpiresAt(this.config.conversationId, renewed.expiresAt);
      this._instance = client;
      sandboxDebugLog("sandbox.renew.success", { instanceId: renewed.instanceId, expiresAt: renewed.expiresAt });
      return renewed;
    } catch (error) {
      sandboxDebugLog("sandbox.renew.failure", { error: summarizeSandboxError(error), instanceId: cached.instanceId });
      if (isRecoverableSandboxError(error)) {
        sandboxDebugLog("sandbox.cache.invalidate", { reason: "renew_error", error: summarizeSandboxError(error), conversationIdPresent: Boolean(this.config.conversationId) });
        this.cache.delete(this.config.conversationId);
        this._instance = null;
      }
      throw error;
    }
  }
  async kill() {
    const cached = this.cache.get(this.config.conversationId);
    if (cached) {
      try {
        const released = await this.apiClient.release(cached.instanceId);
        sandboxDebugLog(released.success ? "sandbox.release.success" : "sandbox.release.failure", { instanceId: cached.instanceId });
      } catch (error) {
        sandboxDebugLog("sandbox.release.failure", { instanceId: cached.instanceId, error: summarizeSandboxError(error) });
      }
    }
    this.cache.delete(this.config.conversationId);
    this._instance = null;
  }
  getInstance() {
    const cached = this.cache.get(this.config.conversationId);
    if (!cached) {
      this._instance = null;
      return null;
    }
    this._instance = cached.sandbox;
    return cached.sandbox;
  }
};
function validateExtendSeconds(seconds) {
  if (!Number.isFinite(seconds) || !Number.isInteger(seconds) || seconds <= 0) {
    throw new Error("extendTimeout seconds must be a positive integer");
  }
}
var RECOVERABLE_SANDBOX_ERROR_CODES = /* @__PURE__ */ new Set([
  "SANDBOX_INSTANCE_UNAVAILABLE",
  "SANDBOX_NETWORK_ERROR",
  "SANDBOX_RESOURCE_NOT_FOUND",
  "SANDBOX_AUTHORIZATION_EXPIRED",
  "SANDBOX_FAILED_OPERATION"
]);
function isRecoverableSandboxError(error) {
  const structuredCode = String(error?.code || "");
  if (RECOVERABLE_SANDBOX_ERROR_CODES.has(structuredCode)) return true;
  const message = String(error?.businessMessage || error?.message || "").toLowerCase();
  const name = String(error?.name || "").toLowerCase();
  const code = String(error?.code || "").toLowerCase();
  const status = Number(error?.status ?? error?.statusCode ?? error?.response?.status);
  const sandboxLifecycleMessage = message.includes("instance not found") || message.includes("sandbox not found") || message.includes("sandbox instance not found") || message.includes("instance not found or not authorized") || message.includes("instance is not running") || message.includes("sandbox unauthorized") || message.includes("unauthorized sandbox") || message.includes("invalid sandbox token") || message.includes("http 404");
  return sandboxLifecycleMessage || name.includes("notfound") || name.includes("unauthorized") || status === 401 || status === 403 || status === 404 || code === "econnrefused" || code === "econnreset" || code === "enotfound" || code === "etimedout";
}
function buildSandboxProxy(config) {
  const lazy = new LazySandbox(config);
  return {
    commands: {
      run: (command, opts) => lazy.withClient((c) => c.commands.run(command, opts))
    },
    files: {
      read: (path) => lazy.withClient((c) => c.files.read(path)),
      write: (path, content) => lazy.withClient((c) => c.files.write(path, content)),
      list: (path) => lazy.withClient((c) => c.files.list(path)),
      makeDir: (path) => lazy.withClient((c) => c.files.makeDir(path)),
      exists: (path) => lazy.withClient((c) => c.files.exists(path)),
      remove: (path) => lazy.withClient((c) => c.files.remove(path))
    },
    browser: {
      get cdpUrl() {
        const instance = lazy.getInstance();
        if (!instance) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
        return instance.browser.cdpUrl;
      },
      get liveUrl() {
        const instance = lazy.getInstance();
        if (!instance) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
        return instance.browser.liveUrl;
      },
      goto: (url, opts) => lazy.withClient((c) => c.browser.goto(url, opts)),
      screenshot: (opts) => lazy.withClient((c) => c.browser.screenshot(opts)),
      click: (selector) => lazy.withClient((c) => c.browser.click(selector)),
      type: (selector, text) => lazy.withClient((c) => c.browser.type(selector, text)),
      evaluate: (script) => lazy.withClient((c) => c.browser.evaluate(script)),
      getContent: () => lazy.withClient((c) => c.browser.getContent()),
      close: () => lazy.withClient((c) => c.browser.close())
    },
    runCode: (code, opts) => lazy.withClient((c) => c.runCode(code, opts)),
    getHost(port) {
      const instance = lazy.getInstance();
      if (!instance) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
      return instance.getHost(port);
    },
    get envdAccessToken() {
      const instance = lazy.getInstance();
      if (!instance) throw new Error("Sandbox is not initialized. Call an async sandbox method first, such as commands.run().");
      return instance.envdAccessToken;
    },
    getInfo: () => lazy.getInfo(),
    extendTimeout: (seconds) => lazy.extendTimeout(seconds),
    kill: () => lazy.kill()
  };
}

// src/tools/websearch/http/client.ts
var import_axios = __toESM(require("axios"), 1);
var SEARCH_PRO_ENDPOINT = "https://api.wsa.cloud.tencent.com/SearchPro";
var DEFAULT_TIMEOUT_MS = 15e3;
var WSA_ERROR_HINTS = {
  InternalError: "Tencent web search internal error. Retry later.",
  InvalidParameter: "Invalid request parameter sent to the web search API.",
  RequestLimitExceeded: "Web search request rate limit exceeded. Retry later.",
  ResourceNotFound: "Web search service is not enabled for this account. Ask the main account to enable Tencent Cloud Web Search (WSA) in the Tencent Cloud console.",
  ResourceUnavailable: "Web search resource is unavailable. Check whether the account is in arrears.",
  UnauthorizedOperation: "Unauthorized web search request. Check that WSA_API_KEY is correct and active."
};
var TencentWsaError = class extends Error {
  constructor(message, opts = { code: "InternalError" }) {
    super(message);
    this.name = "TencentWsaError";
    this.code = opts.code;
    this.requestId = opts.requestId;
    this.status = opts.status;
  }
};
function isDebugEnabled() {
  const env = globalThis.process?.env?.MAKERS_AGENT_TOOLKIT_DEBUG;
  return env === "1" || env === "true";
}
function debugLog(event, data) {
  if (!isDebugEnabled()) return;
  const suffix = data ? ` ${JSON.stringify(data)}` : "";
  console.error(`[pages-agent-toolkit] ${event}${suffix}`);
}
var AxiosWsaTransport = class {
  async post(url, init) {
    const resp = await import_axios.default.request({
      method: "POST",
      url,
      data: init.body,
      headers: init.headers,
      timeout: init.timeoutMs ?? DEFAULT_TIMEOUT_MS,
      responseType: "json",
      transformRequest: [(data) => data],
      validateStatus: () => true
    });
    return { status: resp.status, data: resp.data };
  }
};
function buildAuthError() {
  throw new TencentWsaError(
    "web_search requires the WSA_API_KEY environment variable. Set it to a valid Tencent Cloud Web Search (WSA) API key before calling web_search.",
    { code: "UnauthorizedOperation" }
  );
}
function resolveApiKey() {
  const key = globalThis.process?.env?.WSA_API_KEY;
  if (typeof key !== "string" || key.trim() === "") buildAuthError();
  return key.trim();
}
function parseJson(value) {
  if (value == null) return value;
  if (typeof value === "string") {
    try {
      return JSON.parse(value);
    } catch {
      return value;
    }
  }
  return value;
}
function extractApiError(payload) {
  if (!payload || typeof payload !== "object") return null;
  const root = payload.Response && typeof payload.Response === "object" ? payload.Response : payload;
  const requestId = root.RequestId ?? root.requestId ?? payload.RequestId;
  const error = root.Error ?? root.error;
  if (error && typeof error === "object" && (error.Code ?? error.code)) {
    return {
      code: String(error.Code ?? error.code),
      message: String(error.Message ?? error.message ?? ""),
      requestId
    };
  }
  if (typeof root.Code === "string" && root.Code && root.Pages === void 0) {
    return { code: root.Code, message: String(root.Message ?? ""), requestId };
  }
  return null;
}
function throwApiError(err, status) {
  const hint = WSA_ERROR_HINTS[err.code] ?? "Web search request failed.";
  const detail = err.message ? ` (${err.message})` : "";
  const requestSuffix = err.requestId ? ` [RequestId: ${err.requestId}]` : "";
  throw new TencentWsaError(`web_search failed: ${hint}${detail}${requestSuffix}`, {
    code: err.code,
    requestId: err.requestId,
    status
  });
}
async function searchPro(query, params = {}, transport = new AxiosWsaTransport()) {
  const apiKey = resolveApiKey();
  const requestBody = { Query: query };
  const site = params.site?.trim();
  if (site) requestBody.Site = site;
  const body = JSON.stringify(requestBody);
  debugLog("websearch.wsa.request", { endpoint: SEARCH_PRO_ENDPOINT, site: site || void 0 });
  const { status, data } = await transport.post(SEARCH_PRO_ENDPOINT, {
    headers: {
      "Content-Type": "application/json; charset=UTF-8",
      Authorization: `Bearer ${apiKey}`
    },
    body
  });
  debugLog("websearch.wsa.response", { status });
  const payload = parseJson(data);
  if (status < 200 || status >= 300) {
    const apiError2 = extractApiError(payload);
    if (apiError2) throwApiError(apiError2, status);
    const text = typeof payload === "string" ? payload : JSON.stringify(payload ?? {});
    throw new TencentWsaError(`web_search failed: HTTP ${status} from the web search API. ${text}`.trim(), {
      code: "InternalError",
      status
    });
  }
  const apiError = extractApiError(payload);
  if (apiError) throwApiError(apiError, status);
  const root = payload && payload.Response && typeof payload.Response === "object" ? payload.Response : payload;
  const rawPages = Array.isArray(root?.Pages) ? root.Pages : [];
  const pages = rawPages.map((entry) => parseJson(entry)).filter((entry) => Boolean(entry) && typeof entry === "object");
  return {
    query: root?.Query,
    pages,
    version: root?.Version,
    msg: root?.Msg,
    requestId: root?.RequestId
  };
}

// src/tools/websearch/search.ts
var ENGINE_NAME = "tencent-wsa";
var SimpleSearch = class {
  constructor(transport = new AxiosWsaTransport()) {
    this.transport = transport;
  }
  async search(query, options = {}) {
    if (!query || !query.trim()) {
      throw new Error("query is required");
    }
    const maxResults = options.maxResults ?? 5;
    if (!Number.isInteger(maxResults) || maxResults <= 0) {
      throw new Error("maxResults must be a positive integer");
    }
    const { pages } = await searchPro(query.trim(), { site: options.site }, this.transport);
    return mapPages(pages, maxResults);
  }
};
function mapPages(pages, maxResults) {
  const seen = /* @__PURE__ */ new Set();
  const results = [];
  for (const page of pages) {
    const href = (page.url ?? "").trim();
    if (!href || seen.has(href)) continue;
    seen.add(href);
    results.push({
      title: (page.title ?? "").trim(),
      href,
      snippet: (page.content || page.passage || "").trim(),
      site: (page.site ?? "").trim(),
      date: (page.date ?? "").trim(),
      engine: ENGINE_NAME
    });
    if (results.length >= maxResults) break;
  }
  return results;
}

// src/tools/registry.ts
function getValue(value, key, defaultValue) {
  if (value && typeof value === "object" && key in value) return value[key];
  return defaultValue;
}
function normalizeCommandResult(result) {
  return {
    stdout: getValue(result, "stdout", ""),
    stderr: getValue(result, "stderr", ""),
    exitCode: getValue(result, "exitCode", getValue(result, "exit_code", 0))
  };
}
function normalizeEntry(entry) {
  return {
    name: getValue(entry, "name"),
    type: getValue(entry, "type"),
    path: getValue(entry, "path")
  };
}
function normalizeExecution(execution) {
  const payload = {
    results: getValue(execution, "results", []),
    logs: getValue(execution, "logs", { stdout: [], stderr: [] })
  };
  const error = getValue(execution, "error");
  if (error !== void 0 && error !== null) payload.error = error;
  return payload;
}
function normalizeBrowserResult(result) {
  const payload = { success: getValue(result, "success") };
  const data = getValue(result, "data");
  if (data !== void 0 && data !== null) payload.data = data;
  const error = getValue(result, "error");
  if (error !== void 0 && error !== null) payload.error = error;
  return payload;
}
function asRecord(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}
var BUILTIN_TOOLS = [
  {
    name: "commands",
    description: 'Use this tool when the agent needs to run a one-shot shell command in the sandbox, including builds, tests, package managers, CLI tools, asset downloads, or binary file generation. Prefer files_read or files_write for direct text file reads or writes. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot for page interaction. Prefer code_interpreter for stateful analysis or notebook-style computation. Returns stdout, stderr, and exitCode; exitCode !== 0 means the command failed or completed with an error status. Side effects: shell commands can mutate files, install packages, start processes, or delete data, so destructive commands require extra care. Input example: { "cmd": "npm test", "cwd": "/workspace", "timeout": 60 }. Common mistakes: do not use commands for long-running daemons; do not use shell cat/echo/sed to read or write a single text file when files_read or files_write fits; do not ignore non-zero exitCode. Supports optional cwd, env, and timeout in seconds.',
    inputSchema: {
      type: "object",
      properties: {
        cmd: { type: "string", description: "Shell command to execute as a one-shot sandbox process" },
        cwd: { type: "string", description: "Working directory for the command (optional)" },
        env: { type: "object", description: "Environment variables for this command only (optional)" },
        timeout: { type: "number", description: "Maximum execution time in seconds (optional)" }
      },
      required: ["cmd"]
    },
    execute: async ({ cmd, cwd, env, timeout }, sandbox) => {
      const result = await sandbox.commands.run(cmd, { cwd, env, timeout });
      return normalizeCommandResult(result);
    }
  },
  {
    name: "files_read",
    description: 'Use this tool when the agent needs to read the content of a sandbox text file without running shell commands. Prefer commands for grep/find or shell pipelines, files_list for directory listing, files_write for writing text content, and code_interpreter for executing code or data analysis. Output semantics: returns the file content as a string. Input example: { "path": "/workspace/package.json" }. Common mistakes: do not use files_read for recursive search or shell pipelines; use files_list for directories and files_write for text writes.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file path to read" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => await sandbox.files.read(path)
  },
  {
    name: "files_write",
    description: 'Use this tool when the agent needs to create or overwrite a sandbox text file with exact UTF-8/text content. Prefer commands for generated shell output, pipelines, asset downloads, binary/cache files, or decoding base64 into binary files, and files_read for reading existing text content first. Output semantics: returns { success: true }. Side effects: overwrites existing text file content or creates a new text file. Input example: { "path": "/tmp/out.txt", "content": "hello" }. Common mistakes: writes replace the whole file; content is required; do not use files_write for binary or cache assets such as .mp3, images, archives, or decoded bytes; use files_make_dir for directories and commands for complex generated pipelines.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox text file path to write" },
        content: { type: "string", description: "Exact UTF-8/text content to write; not for binary/cache assets or decoded bytes" }
      },
      required: ["path", "content"]
    },
    execute: async ({ path, content }, sandbox) => {
      if (typeof content !== "string") throw new Error("files_write requires content");
      await sandbox.files.write(path, content);
      return { success: true };
    }
  },
  {
    name: "files_list",
    description: 'Use this tool when the agent needs to list a sandbox directory without running shell commands. Prefer commands for recursive find/grep pipelines. Output semantics: returns entries with name, type, and path only. Input example: { "path": "/workspace" }. Common mistakes: use files_read for file contents; use commands for recursive or filtered searches.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox directory path to list" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      const entries = await sandbox.files.list(path);
      return entries.map((entry) => normalizeEntry(entry));
    }
  },
  {
    name: "files_exists",
    description: 'Use this tool when the agent needs to check whether a sandbox file or directory exists. Output semantics: returns a boolean. Input example: { "path": "/workspace/package.json" }. Common mistakes: files_exists only checks existence; use files_read to inspect content and files_list to inspect directories.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file or directory path to check" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => await sandbox.files.exists(path)
  },
  {
    name: "files_remove",
    description: 'Use this tool when the agent needs to remove a sandbox file or directory. Output semantics: returns { success: true }. Side effects: deletes the target path, so destructive removals require extra care. Input example: { "path": "/tmp/out.txt" }. Common mistakes: do not call files_remove for cleanup unless deletion is intended; use files_exists first when uncertain.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox file or directory path to remove" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      await sandbox.files.remove(path);
      return { success: true };
    }
  },
  {
    name: "files_make_dir",
    description: 'Use this tool when the agent needs to create a sandbox directory, including nested directories supported by the sandbox. Output semantics: returns { success: true }. Side effects: creates directories on the sandbox filesystem. Input example: { "path": "/tmp/new-dir" }. Common mistakes: use files_write for file content; this tool creates directories only and does not accept make_dir or makeDir op values.',
    inputSchema: {
      type: "object",
      properties: {
        path: { type: "string", description: "Sandbox directory path to create" }
      },
      required: ["path"]
    },
    execute: async ({ path }, sandbox) => {
      await sandbox.files.makeDir(path);
      return { success: true };
    }
  },
  {
    name: "browser_fetch",
    description: 'Use this tool when the agent already has a specific target URL and needs a real Chromium browser inside the sandbox to navigate to a page and inspect DOM content at that URL. browser_fetch performs full browser navigation and DOM capture, not a lightweight HTTP/API fetch and not a search tool. Prefer commands with curl or code_interpreter HTTP clients for simple JSON, text, or API requests, and prefer web_search for public discovery, latest news, trending topics, or any information lookup where the exact target URL is not known. Output semantics: returns page metadata plus content truncated to 5000 characters. Side effects: navigation changes the current browser page. Input example: { "url": "https://example.com" }. Common mistakes: do not use browser_fetch for general web search or information discovery; use web_search when you do not already have a specific target URL; do not use browser_fetch for direct API fetches, screenshots, clicks, or form input.',
    inputSchema: {
      type: "object",
      properties: {
        url: { type: "string", description: "Known target URL for a real browser page load and DOM capture; not a lightweight HTTP request and not for search or discovery" }
      },
      required: ["url"]
    },
    execute: async ({ url }, sandbox) => {
      if (typeof url !== "string" || !url) throw new Error("browser_fetch requires url");
      const navigation = await sandbox.browser.goto(url);
      const content = await sandbox.browser.getContent();
      const navigationData = asRecord(getValue(navigation, "data", {}));
      const pageContent = String(getValue(content, "data", "") ?? "");
      return {
        ...navigationData,
        content: pageContent.length > 5e3 ? pageContent.slice(0, 5e3) : pageContent
      };
    }
  },
  {
    name: "browser_screenshot",
    description: 'Use this tool when the agent needs to capture a screenshot from the sandbox browser. Optionally navigates to a URL before capture. Output semantics: returns { base64Image } containing PNG data and does not create a sandbox file or return a file path; the base64Image is the final image payload and can be returned directly to the user as-is. Side effects: optional navigation changes the current browser page. Input example: { "url": "https://example.com", "fullPage": true }. Common mistakes: screenshot returns base64Image, not a file path; do not save the base64Image to a file or convert it before responding unless the user explicitly asks for a file; use browser_fetch for DOM text and browser_click or browser_type for interaction.',
    inputSchema: {
      type: "object",
      properties: {
        url: { type: "string", description: "Optional target URL to navigate before capturing the screenshot" },
        fullPage: { type: "boolean", description: "Capture full page screenshot (default: false). Screenshot output is base64Image, not a file path and does not need to be saved before returning it." }
      },
      required: []
    },
    execute: async ({ url, fullPage }, sandbox) => {
      if (typeof url === "string" && url) await sandbox.browser.goto(url);
      const result = await sandbox.browser.screenshot({ fullPage: Boolean(fullPage) });
      return { base64Image: getValue(result, "base64Image", getValue(result, "base64_image")) };
    }
  },
  {
    name: "browser_click",
    description: 'Use this tool when the agent needs to click an element in the current sandbox browser page. Output semantics: returns normalized success/data/error payloads. Side effects: clicking can change page state. Input example: { "selector": "#submit" }. Common mistakes: browser_click requires a CSS selector and depends on the current page state; use browser_fetch first if navigation or DOM capture is needed.',
    inputSchema: {
      type: "object",
      properties: {
        selector: { type: "string", description: "CSS selector of the element to click" }
      },
      required: ["selector"]
    },
    execute: async ({ selector }, sandbox) => {
      if (typeof selector !== "string" || !selector) throw new Error("browser_click requires selector");
      const result = await sandbox.browser.click(selector);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "browser_type",
    description: 'Use this tool when the agent needs to type text into an element in the current sandbox browser page. Output semantics: returns normalized success/data/error payloads. Side effects: typing can change page state. Input example: { "selector": "#name", "text": "Alice" }. Common mistakes: browser_type requires both selector and text and depends on the current page state; it does not submit forms unless typing itself triggers submission.',
    inputSchema: {
      type: "object",
      properties: {
        selector: { type: "string", description: "CSS selector of the input element to type into" },
        text: { type: "string", description: "Text to type into the selected element" }
      },
      required: ["selector", "text"]
    },
    execute: async ({ selector, text }, sandbox) => {
      if (typeof selector !== "string" || !selector || typeof text !== "string") throw new Error("browser_type requires selector and text");
      const result = await sandbox.browser.type(selector, text);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "browser_evaluate",
    description: 'Use this tool when the agent needs to evaluate JavaScript in the current sandbox browser page context. Output semantics: returns normalized success/data/error payloads. Side effects: evaluated scripts can change page state. Input example: { "script": "document.title" }. Common mistakes: browser_evaluate runs in the current page and can mutate it; use code_interpreter for sandbox computation and commands for CLI workflows.',
    inputSchema: {
      type: "object",
      properties: {
        script: { type: "string", description: "JavaScript code to evaluate in the page context" }
      },
      required: ["script"]
    },
    execute: async ({ script }, sandbox) => {
      if (typeof script !== "string" || !script) throw new Error("browser_evaluate requires script");
      const result = await sandbox.browser.evaluate(script);
      return normalizeBrowserResult(result);
    }
  },
  {
    name: "code_interpreter",
    description: 'Use this tool when the agent needs to execute code in a stateful Jupyter kernel inside the sandbox for data analysis, calculations, notebooks, and exploratory code. Variables and imports persist across calls, making it useful for iterative computation. Prefer commands for project builds, tests, package managers, and long-running shell workflows. Prefer files_read or files_write for direct text file edits. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot for real page interaction. Returns results, logs.stdout, logs.stderr, and error when execution fails. Side effects: kernel state persists across calls and can affect later executions. Input example: { "language": "python", "code": "import statistics\\nstatistics.mean([1, 2, 3])", "timeout": 30 }. Common mistakes: do not use code_interpreter for project tests or package-manager commands; do not assume a clean kernel; do not treat it as a stateless executor; use commands for CLI workflows. Supports optional timeout in seconds.',
    inputSchema: {
      type: "object",
      properties: {
        language: { type: "string", enum: ["python", "javascript", "r", "bash"], description: "Kernel language to execute: python, javascript, r, or bash" },
        code: { type: "string", description: "Code to execute in the selected stateful kernel" },
        timeout: { type: "number", description: "Maximum execution time in seconds for this code cell (optional)" }
      },
      required: ["language", "code"]
    },
    execute: async ({ language, code, timeout }, sandbox) => {
      const result = await sandbox.runCode(code, { language, timeout });
      return normalizeExecution(result);
    }
  },
  {
    // node-websearch-tool.REGISTRY.1 node-websearch-tool.SCOPE.2 node-websearch-tool.SCOPE.3
    name: "web_search",
    description: 'Use this tool when the agent needs lightweight web search results for current facts, documentation, references, or discovery. Use web_search as the first choice to find, discover, or look up information from the web, including latest news, trending topics, real-time events, documentation, references, or any query where a specific URL is not already known. It is backed by the Tencent Cloud web search (WSA) API and requires the WSA_API_KEY environment variable. Prefer browser_fetch, browser_click, browser_type, browser_evaluate, or browser_screenshot when the agent needs real page interaction, DOM inspection, screenshots, form entry, or JavaScript execution; browser_fetch performs real Chromium navigation and DOM capture, not general search. Prefer commands or code_interpreter for direct HTTP/API requests, local computation, project builds, or sandbox file work. Returns an array of results with title, href, and snippet, plus the source site and publication date when available. Side effects: performs outbound web requests only and does not use or mutate the sandbox filesystem, shell, browser, or kernel state. Input example: { "query": "EdgeOne Pages documentation", "maxResults": 5 }. Common mistakes: do not use web_search for interacting with a specific page; do not use it for direct JSON/API fetching; do not expect browser session state, screenshots, or JavaScript execution; do not call browser_fetch first for open-ended searches or latest-news queries without a known URL. Supports an optional maxResults limit and an optional site for site-restricted search.',
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string", description: "Search query to send to the web search API. Required and must not be empty. Use this when the target URL is unknown." },
        maxResults: { type: "integer", minimum: 1, description: "Maximum number of deduplicated search results to return (optional, default 5). Must be a positive integer." },
        site: { type: "string", description: 'Optional single domain to restrict results to (site-restricted search), e.g. "zhihu.com". Omit to search the whole web.' }
      },
      required: ["query"]
    },
    execute: async ({ query, maxResults, site }) => {
      const search = new SimpleSearch();
      const options = { maxResults };
      if (site !== void 0 && site !== null) options.site = site;
      return search.search(query, options);
    }
  }
];

// src/tools/errors.ts
function isToolErrorResult(value) {
  if (!value || typeof value !== "object") return false;
  const record = value;
  if (record.isError !== true) return false;
  const error = record.error;
  return !!error && typeof error === "object" && typeof error.message === "string";
}
function toToolErrorResult(toolName, error) {
  const message = error instanceof Error ? error.message : String(error);
  const type = error instanceof Error && error.name ? error.name : "Error";
  return { isError: true, error: { tool: toolName, type, message } };
}

// src/tools/adapters/claude.ts
function wrapForClaude(meta, sandbox) {
  return {
    name: meta.name,
    description: meta.description,
    input_schema: meta.inputSchema,
    async execute(args) {
      return await meta.execute(args, sandbox);
    }
  };
}

// src/tools/adapters/openai.ts
function wrapForOpenAI(meta, sandbox) {
  return {
    type: "function",
    function: {
      name: meta.name,
      description: meta.description,
      parameters: meta.inputSchema
    },
    async execute(args) {
      return await meta.execute(args, sandbox);
    }
  };
}

// src/tools/adapters/openai-strict-schema.ts
function stripNullishArgs(args) {
  const normalized = {};
  for (const [key, value] of Object.entries(args || {})) {
    if (value !== null && value !== void 0) normalized[key] = value;
  }
  return normalized;
}
function withDescription(base, propertySchema) {
  void propertySchema;
  return base;
}
function propertyToOpenAIStrictSchema(propertySchema) {
  if (Array.isArray(propertySchema?.enum)) {
    const values = propertySchema.enum.filter((item) => typeof item === "string");
    return withDescription(
      values.length > 0 ? { type: "string", enum: values } : { type: "string" },
      propertySchema
    );
  }
  const type = Array.isArray(propertySchema?.type) ? propertySchema.type.find((item) => item !== "null") : propertySchema?.type;
  if (type === "number" || type === "integer") return withDescription({ type: "number" }, propertySchema);
  if (type === "boolean") return withDescription({ type: "boolean" }, propertySchema);
  if (type === "object") {
    return withDescription({ type: "object", additionalProperties: {} }, propertySchema);
  }
  if (type === "array") return withDescription({ type: "array", items: {} }, propertySchema);
  return withDescription({ type: "string" }, propertySchema);
}
function toNullableSchema(propertySchema) {
  return {
    anyOf: [propertyToOpenAIStrictSchema(propertySchema), { type: "null" }]
  };
}
function jsonSchemaToOpenAIStrictSchema(schema) {
  const properties = schema?.properties && typeof schema.properties === "object" ? schema.properties : {};
  const required = Array.isArray(schema?.required) ? new Set(schema.required) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(properties)) {
    shape[key] = required.has(key) ? propertyToOpenAIStrictSchema(propertySchema) : toNullableSchema(propertySchema);
  }
  return {
    type: "object",
    properties: shape,
    required: Object.keys(properties),
    additionalProperties: false
  };
}

// src/tools/adapters/openai-agents.ts
function parseToolInput(payload) {
  if (typeof payload === "string") {
    try {
      const parsed = JSON.parse(payload);
      return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : {};
    } catch {
      return {};
    }
  }
  return payload && typeof payload === "object" && !Array.isArray(payload) ? payload : {};
}
function wrapForOpenAIAgents(meta, sandbox) {
  const execute = async (args) => {
    return await meta.execute(stripNullishArgs(args), sandbox);
  };
  return {
    type: "function",
    name: meta.name,
    description: meta.description,
    strict: true,
    parameters: jsonSchemaToOpenAIStrictSchema(meta.inputSchema),
    async execute(args) {
      return await execute(args);
    },
    async invoke(contextOrInput, input) {
      const payload = input === void 0 ? contextOrInput : input;
      return await execute(parseToolInput(payload));
    },
    async needsApproval() {
      return false;
    },
    async isEnabled() {
      return true;
    }
  };
}

// src/tools/adapters/framework-helpers.ts
var import_zod = require("zod");
function stringifyToolResult(result) {
  if (typeof result === "string") return result;
  try {
    return JSON.stringify(result);
  } catch {
    return String(result);
  }
}
function stripNullishArgs2(args) {
  return Object.fromEntries(
    Object.entries(args || {}).filter(([, value]) => value !== null && value !== void 0)
  );
}
function primitiveSchemaToZod(schema) {
  const enumValues = Array.isArray(schema?.enum) ? schema.enum.filter((value) => typeof value === "string") : [];
  if (enumValues.length > 0) {
    return import_zod.z.enum(enumValues);
  }
  switch (schema?.type) {
    case "string":
      return import_zod.z.string();
    case "number":
      return import_zod.z.number();
    case "integer":
      return import_zod.z.number().int();
    case "boolean":
      return import_zod.z.boolean();
    case "array":
      return import_zod.z.array(jsonSchemaToZod(schema.items || {}));
    case "object":
      return jsonSchemaToZod(schema);
    default:
      return import_zod.z.any();
  }
}
function withDescription2(zodSchema, schema) {
  return typeof schema?.description === "string" && schema.description ? zodSchema.describe(schema.description) : zodSchema;
}
function jsonSchemaToZod(schema) {
  if (schema?.type !== "object") {
    return withDescription2(primitiveSchemaToZod(schema), schema);
  }
  if (!schema?.properties || typeof schema.properties !== "object") {
    return withDescription2(import_zod.z.record(import_zod.z.string(), import_zod.z.any()), schema);
  }
  const required = Array.isArray(schema.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, value] of Object.entries(schema.properties)) {
    const fieldSchema = withDescription2(primitiveSchemaToZod(value), value);
    shape[key] = required.has(key) ? fieldSchema : fieldSchema.nullish();
  }
  return import_zod.z.object(shape);
}
function bindFrameworkHandler(meta, sandbox) {
  return async (args) => await meta.execute(stripNullishArgs2(args), sandbox);
}
function buildLangChainTool(meta, sandbox, toolFactory) {
  const handler = bindFrameworkHandler(meta, sandbox);
  const wrappedTool = toolFactory(
    async (args) => stringifyToolResult(await handler(args)),
    {
      name: meta.name,
      description: meta.description,
      schema: jsonSchemaToZod(meta.inputSchema)
    }
  );
  return Object.assign(wrappedTool, { handler });
}
function isClassFactory(value) {
  return typeof value === "function" && /^class\s/.test(Function.prototype.toString.call(value));
}
function buildCrewAITool(meta, sandbox, baseTool) {
  const handler = bindFrameworkHandler(meta, sandbox);
  const coroutine = async (args) => stringifyToolResult(await handler(args));
  const config = {
    name: meta.name,
    description: meta.description,
    args_schema: jsonSchemaToZod(meta.inputSchema)
  };
  let crewTool;
  if (isClassFactory(baseTool)) {
    class SandboxCrewAITool extends baseTool {
      async _run(args = {}) {
        return await coroutine(args);
      }
      async _arun(args = {}) {
        return await coroutine(args);
      }
    }
    crewTool = new SandboxCrewAITool(config);
  } else if (typeof baseTool === "function") {
    const factory = baseTool;
    crewTool = factory({
      ...config,
      coroutine,
      handler
    });
    if (typeof crewTool._run !== "function") {
      crewTool._run = async (args = {}) => await coroutine(args);
    }
    if (typeof crewTool._arun !== "function") {
      crewTool._arun = async (args = {}) => await coroutine(args);
    }
  } else {
    throw new TypeError("baseTool must be a CrewAI-compatible class or factory");
  }
  return Object.assign(crewTool, { handler });
}

// src/tools/adapters/langgraph.ts
function wrapForLangGraph(meta, sandbox) {
  return buildLangChainTool(meta, sandbox, (handler, config) => ({
    ...config,
    invoke: handler
  }));
}

// src/tools/adapters/crewai.ts
var LightweightCrewAITool = class {
  constructor(config) {
    this.name = config.name;
    this.description = config.description;
    this.args_schema = config.args_schema;
  }
};
function wrapForCrewAI(meta, sandbox) {
  return buildCrewAITool(meta, sandbox, LightweightCrewAITool);
}

// src/tools/adapters/deepagents.ts
function wrapForDeepAgents(meta, sandbox) {
  return buildLangChainTool(meta, sandbox, (handler, config) => ({
    ...config,
    invoke: handler,
    call: handler
  }));
}

// src/tools/adapters/claude-mcp.ts
var import_zod2 = require("zod");
var DEFAULT_CLAUDE_MCP_SERVER_NAME = "edgeone";
function normalizeServerName(name) {
  const normalized = typeof name === "string" ? name.trim() : "";
  return normalized || DEFAULT_CLAUDE_MCP_SERVER_NAME;
}
function withDescription3(zodSchema, schema) {
  return typeof schema?.description === "string" && schema.description ? zodSchema.describe(schema.description) : zodSchema;
}
function schemaToZod(schema) {
  const enumValues = Array.isArray(schema?.enum) ? schema.enum.filter((value) => typeof value === "string") : [];
  if (enumValues.length > 0) {
    return withDescription3(import_zod2.z.enum(enumValues), schema);
  }
  const type = Array.isArray(schema?.type) ? schema.type.find((item) => item !== "null") : schema?.type;
  let result;
  switch (type) {
    case "number":
      result = import_zod2.z.number();
      break;
    case "integer":
      result = import_zod2.z.number().int();
      break;
    case "boolean":
      result = import_zod2.z.boolean();
      break;
    case "array":
      result = import_zod2.z.array(schemaToZod(schema.items || {}));
      break;
    case "object":
      result = objectSchemaToZod(schema);
      break;
    case "string":
      result = import_zod2.z.string();
      break;
    default:
      result = import_zod2.z.any();
  }
  if (Array.isArray(schema?.type) && schema.type.includes("null")) {
    result = result.nullable();
  }
  return withDescription3(result, schema);
}
function objectSchemaToZod(schema) {
  if (!schema?.properties || typeof schema.properties !== "object") {
    return import_zod2.z.record(import_zod2.z.string(), import_zod2.z.any());
  }
  const required = Array.isArray(schema?.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(schema.properties)) {
    const field = schemaToZod(propertySchema);
    shape[key] = required.has(key) ? field : field.optional();
  }
  return import_zod2.z.object(shape);
}
function jsonSchemaToClaudeInputShape(schema) {
  if (!schema || schema.type !== "object" || !schema.properties || typeof schema.properties !== "object") {
    return void 0;
  }
  const required = Array.isArray(schema.required) ? new Set(schema.required.filter((key) => typeof key === "string")) : /* @__PURE__ */ new Set();
  const shape = {};
  for (const [key, propertySchema] of Object.entries(schema.properties)) {
    const field = schemaToZod(propertySchema);
    shape[key] = required.has(key) ? field : field.optional();
  }
  return shape;
}
function stringifyToolResult2(result) {
  if (typeof result === "string") return result;
  const json = JSON.stringify(result, null, 2);
  return typeof json === "string" ? json : String(result);
}
function toolToClaudeMcp(meta, sandbox, options) {
  if (!meta || typeof meta.name !== "string" || !meta.name || typeof meta.description !== "string" || typeof meta.execute !== "function") {
    return void 0;
  }
  const inputSchema = jsonSchemaToClaudeInputShape(meta.inputSchema);
  if (!inputSchema) return void 0;
  const tool = {
    name: meta.name,
    description: meta.description,
    inputSchema,
    async handler(args) {
      try {
        const result = await meta.execute(args, sandbox);
        if (isToolErrorResult(result)) {
          return { content: [{ type: "text", text: result.error.message }], isError: true };
        }
        return { content: [{ type: "text", text: stringifyToolResult2(result) }] };
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return { content: [{ type: "text", text: message }], isError: true };
      }
    }
  };
  if (options?.alwaysLoad ?? true) {
    tool._meta = { "anthropic/alwaysLoad": true };
  }
  return tool;
}
function buildClaudeMcpServerBundle(metadata, sandbox, name, options) {
  const serverName = normalizeServerName(name);
  const tools = metadata.map((meta) => toolToClaudeMcp(meta, sandbox, options)).filter((tool) => Boolean(tool));
  return {
    name: serverName,
    tools,
    allowedTools: tools.map((tool) => `mcp__${serverName}__${tool.name}`)
  };
}

// src/tools/index.ts
var FILE_TOOL_NAMES = [
  "files_read",
  "files_write",
  "files_list",
  "files_exists",
  "files_remove",
  "files_make_dir"
];
var BROWSER_TOOL_NAMES = [
  "browser_fetch",
  "browser_screenshot",
  "browser_click",
  "browser_type",
  "browser_evaluate"
];
function wrapForFramework(framework, meta, sandbox) {
  switch (framework) {
    case "claude-agent-sdk":
      return wrapForClaude(meta, sandbox);
    case "openai-sdk":
      return wrapForOpenAI(meta, sandbox);
    case "openai-agents-sdk":
      return wrapForOpenAIAgents(meta, sandbox);
    case "langgraph":
      return wrapForLangGraph(meta, sandbox);
    case "crewai":
      return wrapForCrewAI(meta, sandbox);
    case "deepagents":
      return wrapForDeepAgents(meta, sandbox);
    case "basic":
    default:
      return wrapForClaude(meta, sandbox);
  }
}
function isDebugEnabled2(options) {
  const env = globalThis.process?.env?.MAKERS_AGENT_TOOLKIT_DEBUG;
  return options?.debug === true || env === "1" || env === "true";
}
function sanitize(value) {
  if (Array.isArray(value)) return value.map(sanitize);
  if (!value || typeof value !== "object") return value;
  return Object.fromEntries(Object.entries(value).map(([key, item]) => {
    const lowered = key.toLowerCase();
    if (lowered.includes("token") || lowered.includes("auth") || lowered.includes("password") || lowered.includes("secret") || lowered.includes("key")) {
      return [key, "[redacted]"];
    }
    if (key === "base64Image" && typeof item === "string") {
      return ["base64ImageLength", item.length];
    }
    return [key, sanitize(item)];
  }));
}
function summarizeResult(result) {
  if (result && typeof result === "object" && typeof result.base64Image === "string") {
    return { base64ImageLength: result.base64Image.length, savedToFile: false };
  }
  return sanitize(result);
}
function debugLog2(enabled, event, data) {
  if (!enabled) return;
  const suffix = data ? ` ${JSON.stringify(sanitize(data))}` : "";
  console.error(`[pages-agent-toolkit] ${event}${suffix}`);
}
function webSearchDebugFields(toolName, args, result) {
  if (toolName !== "web_search" || !args || typeof args !== "object") return {};
  const record = args;
  const fields = {};
  if (typeof record.query === "string") fields.query = record.query;
  if (typeof record.site === "string" && record.site) fields.site = record.site;
  if (record.maxResults !== void 0 && record.maxResults !== null) fields.maxResults = record.maxResults;
  if (Array.isArray(result)) fields.resultCount = result.length;
  return fields;
}
function stripWebSearchEngine(toolName, result) {
  if (toolName !== "web_search" || !Array.isArray(result)) return result;
  return result.map((item) => {
    if (!item || typeof item !== "object" || Array.isArray(item)) return item;
    const { engine: _engine, ...rest } = item;
    return rest;
  });
}
function withDebug(meta, enabled) {
  if (!enabled) {
    if (meta.name === "web_search") {
      return {
        ...meta,
        async execute(args, sandbox) {
          const result = await meta.execute(args, sandbox);
          return stripWebSearchEngine(meta.name, result);
        }
      };
    }
    return meta;
  }
  return {
    ...meta,
    async execute(args, sandbox) {
      const operation = meta.name;
      const start = Date.now();
      const preFields = webSearchDebugFields(meta.name, args);
      debugLog2(true, "tool.execute.start", { tool: meta.name, operation, ...preFields, args });
      try {
        const result = await meta.execute(args, sandbox);
        const postFields = webSearchDebugFields(meta.name, args, result);
        debugLog2(true, "tool.execute.success", {
          tool: meta.name,
          operation,
          ...postFields,
          durationMs: Date.now() - start,
          result: summarizeResult(result)
        });
        return stripWebSearchEngine(meta.name, result);
      } catch (error) {
        debugLog2(true, "tool.execute.error", {
          tool: meta.name,
          operation,
          ...preFields,
          durationMs: Date.now() - start,
          error: error?.message || String(error)
        });
        throw error;
      }
    }
  };
}
function withErrorHandling(meta) {
  return {
    ...meta,
    async execute(args, sandbox) {
      try {
        return await meta.execute(args, sandbox);
      } catch (error) {
        return toToolErrorResult(meta.name, error);
      }
    }
  };
}
function pickMetadata(metadata, names) {
  if (names === void 0) return metadata;
  const requestedNames = typeof names === "string" ? [names] : names;
  const seen = /* @__PURE__ */ new Set();
  const selected = [];
  for (const name of requestedNames) {
    if (seen.has(name)) continue;
    const meta = metadata.find((item) => item.name === name);
    if (!meta) throw new Error(`Unknown tool name: ${name}`);
    seen.add(name);
    selected.push(meta);
  }
  return selected;
}
function buildTools(framework, sandbox, options) {
  const fw = framework || "claude-agent-sdk";
  const toolMap = {};
  const debugEnabled = isDebugEnabled2(options);
  const metadata = BUILTIN_TOOLS.map((meta) => withErrorHandling(withDebug(meta, debugEnabled)));
  for (const meta of metadata) {
    toolMap[meta.name] = wrapForFramework(fw, meta, sandbox);
  }
  const pickTools = (names) => names.map((name) => toolMap[name]).filter(Boolean);
  return {
    commands: toolMap.commands,
    files_read: toolMap.files_read,
    files_write: toolMap.files_write,
    files_list: toolMap.files_list,
    files_exists: toolMap.files_exists,
    files_remove: toolMap.files_remove,
    files_make_dir: toolMap.files_make_dir,
    browser_fetch: toolMap.browser_fetch,
    browser_screenshot: toolMap.browser_screenshot,
    browser_click: toolMap.browser_click,
    browser_type: toolMap.browser_type,
    browser_evaluate: toolMap.browser_evaluate,
    code_interpreter: toolMap.code_interpreter,
    // node-websearch-tool.REGISTRY.2
    web_search: toolMap.web_search,
    all: () => Object.values(toolMap),
    get: (name) => toolMap[name],
    files: () => pickTools(FILE_TOOL_NAMES),
    browser: () => pickTools(BROWSER_TOOL_NAMES),
    // node-websearch-tool.COMPATIBILITY.3
    toLangChainTools: (toolFactory, names) => pickMetadata(metadata, names).map((meta) => buildLangChainTool(meta, sandbox, toolFactory)),
    // node-websearch-tool.COMPATIBILITY.4
    toCrewAITools: (baseTool, names) => pickMetadata(metadata, names).map((meta) => buildCrewAITool(meta, sandbox, baseTool)),
    toClaudeMcpServer: (name, options2) => buildClaudeMcpServerBundle(metadata, sandbox, name, options2)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SandboxNotAvailableError,
  buildSandboxProxy,
  buildTools,
  isToolErrorResult,
  toToolErrorResult
});
