// @edgeone/nuxt-pages/dist/esm-chunks/chunk-ZR3DYEBK.js

      var require = await (async () => {
        var { createRequire } = await import("node:module");
        return createRequire(import.meta.url);
      })();
    
import {
  require_lib
} from "./chunk-VCJ5U4PV.js";
import {
  __toESM
} from "./chunk-6BT4RYQJ.js";

// src/build/content/module-inject.ts
var parser = __toESM(require_lib(), 1);
import { existsSync } from "node:fs";
import { readFile, writeFile, rm } from "node:fs/promises";
import { join } from "node:path";
import { createRequire } from "node:module";
var require2 = createRequire(import.meta.url);
var traverse = require2("@babel/traverse").default;
var generate = require2("@babel/generator").default;
var TEMP_MODULE_FILENAME = ".edgeone-nitro-config.mjs";
function generateModuleContent() {
  return `// Auto-generated by @edgeone/nuxt-pages \u2014 DO NOT EDIT
// This file will be removed after build completes.
import { defineNuxtModule } from '@nuxt/kit'

export default defineNuxtModule({
  meta: { name: 'edgeone-nitro-config' },
  setup(_options, nuxt) {
    nuxt.hook('nitro:config', (nitroConfig) => {
      // Store original values for reference
      const originalOutput = nitroConfig.output ? { ...nitroConfig.output } : null
      const originalPreset = nitroConfig.preset

      // Override output directories for EdgeOne deployment
      nitroConfig.output = nitroConfig.output || {}
      nitroConfig.output.dir = '.edgeone'
      nitroConfig.output.publicDir = '.edgeone/assets'
      nitroConfig.output.serverDir = '.edgeone/cloud-functions/ssr-node'

      // Remove preset to use default node-server
      if (nitroConfig.preset && nitroConfig.preset !== 'node-server') {
        delete nitroConfig.preset
      }
    })
  }
})
`;
}
async function writeTempModule(projectRoot) {
  const modulePath = join(projectRoot, TEMP_MODULE_FILENAME);
  await writeFile(modulePath, generateModuleContent(), "utf-8");
  console.log(`Created temporary EdgeOne module: ${modulePath}`);
  return modulePath;
}
async function removeTempModule(projectRoot) {
  const modulePath = join(projectRoot, TEMP_MODULE_FILENAME);
  if (existsSync(modulePath)) {
    await rm(modulePath, { force: true });
    console.log(`Removed temporary EdgeOne module: ${modulePath}`);
  }
}
function addModuleToConfig(code, moduleRef) {
  try {
    const ast = parser.parse(code, {
      sourceType: "module",
      plugins: ["typescript"]
    });
    let configObjectNode = null;
    let modulesArrayNode = null;
    traverse(ast, {
      CallExpression(path) {
        const { node } = path;
        if (node.callee.type === "Identifier" && node.callee.name === "defineNuxtConfig" && node.arguments.length > 0) {
          let arg = node.arguments[0];
          while (arg.type === "TSAsExpression" || arg.type === "TSSatisfiesExpression" || arg.type === "TSTypeAssertion") {
            arg = arg.expression;
          }
          if (arg.type !== "ObjectExpression") return;
          configObjectNode = arg;
          for (const prop of configObjectNode.properties) {
            if (prop.type === "ObjectProperty" && prop.key.type === "Identifier" && prop.key.name === "modules" && prop.value.type === "ArrayExpression") {
              modulesArrayNode = prop.value;
              break;
            }
          }
          path.stop();
        }
      }
    });
    if (!configObjectNode) {
      return null;
    }
    const originalCode = code;
    if (modulesArrayNode) {
      const stringLiteral = {
        type: "StringLiteral",
        value: moduleRef
      };
      modulesArrayNode.elements.unshift(stringLiteral);
    } else {
      const modulesProp = {
        type: "ObjectProperty",
        key: { type: "Identifier", name: "modules" },
        value: {
          type: "ArrayExpression",
          elements: [{ type: "StringLiteral", value: moduleRef }]
        },
        computed: false,
        shorthand: false
      };
      configObjectNode.properties.unshift(modulesProp);
    }
    const output = generate(ast, { retainLines: true }, code);
    return { newCode: output.code, originalCode };
  } catch (e) {
    console.error("Failed to inject module into nuxt config:", e);
    return null;
  }
}
async function injectEdgeoneModule(configPath, projectRoot) {
  const result = {
    success: false,
    originalConfigContent: null,
    configPath,
    projectRoot
  };
  try {
    await writeTempModule(projectRoot);
    if (!configPath || !existsSync(configPath)) {
      const defaultConfig = `export default defineNuxtConfig({
  modules: ['./${TEMP_MODULE_FILENAME}'],
})`;
      const newConfigPath = join(projectRoot, "nuxt.config.ts");
      await writeFile(newConfigPath, defaultConfig, "utf-8");
      console.log(`Created nuxt.config.ts with EdgeOne module at: ${newConfigPath}`);
      result.configPath = newConfigPath;
      result.originalConfigContent = null;
      result.success = true;
      return result;
    }
    const configContent = await readFile(configPath, "utf-8");
    const moduleRef = `./${TEMP_MODULE_FILENAME}`;
    const astResult = addModuleToConfig(configContent, moduleRef);
    if (astResult) {
      await writeFile(configPath, astResult.newCode, "utf-8");
      console.log(`Injected EdgeOne module into: ${configPath}`);
      result.originalConfigContent = astResult.originalCode;
      result.success = true;
    } else {
      console.error("Failed to inject module via AST, cleaning up temp module file");
      await removeTempModule(projectRoot);
    }
  } catch (error) {
    console.error("injectEdgeoneModule failed:", error);
    await removeTempModule(projectRoot).catch(() => {
    });
  }
  return result;
}
async function cleanupEdgeoneModule(injectionResult) {
  try {
    const { originalConfigContent, configPath, projectRoot } = injectionResult;
    if (originalConfigContent !== null && existsSync(configPath)) {
      await writeFile(configPath, originalConfigContent, "utf-8");
      console.log(`Restored nuxt config: ${configPath}`);
    }
    await removeTempModule(projectRoot);
  } catch (error) {
    console.error("cleanupEdgeoneModule failed:", error);
  }
}

export {
  removeTempModule,
  injectEdgeoneModule,
  cleanupEdgeoneModule
};
